{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "file = '/home/grads/z/zshuying/Documents/shuying/ppi_mutation/data/clinvar/Jan19.txt'\n",
    "out='/home/grads/z/zshuying/Documents/shuying/ppi_mutation/data/clinvar/criteria_above_discard_redundant_clinsig.txt'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{' other'}  is discarded\n",
      "{' other'}  is discarded\n",
      "{' other'}  is discarded\n",
      "{' other'}  is discarded\n",
      "{' other'}  is discarded\n",
      "{' other'}  is discarded\n",
      "{' other'}  is discarded\n",
      "{' other'}  is discarded\n",
      "{' risk factor'}  is discarded\n",
      "{' risk factor'}  is discarded\n",
      "{' risk factor'}  is discarded\n",
      "{' risk factor'}  is discarded\n",
      "{' risk factor'}  is discarded\n",
      "{' other'}  is discarded\n",
      "{' other'}  is discarded\n",
      "{' other'}  is discarded\n",
      "{' risk factor'}  is discarded\n",
      "{' association', ' risk factor'}  is discarded\n",
      "{' association'}  is discarded\n",
      "{' other'}  is discarded\n",
      "{' other'}  is discarded\n",
      "{' risk factor'}  is discarded\n",
      "{' other'}  is discarded\n",
      "{' other'}  is discarded\n",
      "{' risk factor'}  is discarded\n",
      "{' risk factor'}  is discarded\n",
      "{' risk factor'}  is discarded\n",
      "{' association'}  is discarded\n",
      "{' risk factor', ' other'}  is discarded\n",
      "{' other'}  is discarded\n",
      "{' other'}  is discarded\n",
      "{' other'}  is discarded\n",
      "{' other'}  is discarded\n",
      "{' other'}  is discarded\n",
      "{' other'}  is discarded\n",
      "{' other'}  is discarded\n",
      "{' other'}  is discarded\n",
      "{' other'}  is discarded\n",
      "{' other'}  is discarded\n",
      "{' risk factor'}  is discarded\n",
      "{' other'}  is discarded\n",
      "{' risk factor'}  is discarded\n",
      "{' risk factor'}  is discarded\n",
      "{' risk factor'}  is discarded\n",
      "{' association'}  is discarded\n",
      "{' risk factor'}  is discarded\n"
     ]
    }
   ],
   "source": [
    "with open(file,'r') as f:\n",
    "    with open(out,'w') as f_out:\n",
    "        line=f.readline()\n",
    "        f_out.write('clinvar_id;review_status;clinical_sig;uniprot_kb;variant_type;hgvs_p;missense\\n')\n",
    "        while True:\n",
    "            line=f.readline()\n",
    "\n",
    "            if line.count(';')>6:\n",
    "                discard_num=line.count(';')-6+3\n",
    "                line_new=';'.join(line.split(';')[:3]+line.split(';')[discard_num:])\n",
    "                print(set(line.split(';'))-set(line_new.split(';')),' is discarded')\n",
    "                f_out.write(line_new)\n",
    "            else:\n",
    "                f_out.write(line)\n",
    "            if not line:break"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Inspect data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>clinvar_id</th>\n",
       "      <th>review_status</th>\n",
       "      <th>clinical_sig</th>\n",
       "      <th>uniprot_kb</th>\n",
       "      <th>variant_type</th>\n",
       "      <th>hgvs_p</th>\n",
       "      <th>missense</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SCV002232601</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>P04181#VAR_000578</td>\n",
       "      <td>Variation</td>\n",
       "      <td>P04181:p.Arg271Lys</td>\n",
       "      <td>missense variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SCV002291827</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>P04181#VAR_015648</td>\n",
       "      <td>Variation</td>\n",
       "      <td>P04181:p.Gln90Glu</td>\n",
       "      <td>missense variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SCV001445903</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>Likely pathogenic</td>\n",
       "      <td>P12955#VAR_011615</td>\n",
       "      <td>Variation</td>\n",
       "      <td>P12955:p.Gly278Asp</td>\n",
       "      <td>missense variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SCV001449002</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>Pathogenic/Likely pathogenic</td>\n",
       "      <td>P12955#VAR_004405</td>\n",
       "      <td>Variation</td>\n",
       "      <td>P12955:p.Gly448Arg</td>\n",
       "      <td>missense variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SCV000935565</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>Likely pathogenic</td>\n",
       "      <td>P21359#VAR_002663</td>\n",
       "      <td>Variation</td>\n",
       "      <td>P21359:p.Leu1953Pro</td>\n",
       "      <td>missense variant</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     clinvar_id                        review_status  \\\n",
       "0  SCV002232601  criteria provided, single submitter   \n",
       "1  SCV002291827  criteria provided, single submitter   \n",
       "2  SCV001445903  criteria provided, single submitter   \n",
       "3  SCV001449002  criteria provided, single submitter   \n",
       "4  SCV000935565  criteria provided, single submitter   \n",
       "\n",
       "                   clinical_sig         uniprot_kb variant_type  \\\n",
       "0                    Pathogenic  P04181#VAR_000578    Variation   \n",
       "1        Uncertain significance  P04181#VAR_015648    Variation   \n",
       "2             Likely pathogenic  P12955#VAR_011615    Variation   \n",
       "3  Pathogenic/Likely pathogenic  P12955#VAR_004405    Variation   \n",
       "4             Likely pathogenic  P21359#VAR_002663    Variation   \n",
       "\n",
       "                hgvs_p          missense  \n",
       "0   P04181:p.Arg271Lys  missense variant  \n",
       "1    P04181:p.Gln90Glu  missense variant  \n",
       "2   P12955:p.Gly278Asp  missense variant  \n",
       "3   P12955:p.Gly448Arg  missense variant  \n",
       "4  P21359:p.Leu1953Pro  missense variant  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data=pd.read_csv(out,delimiter=';')\n",
    "data.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Pathogenic', 'Uncertain significance', 'Likely pathogenic',\n",
       "       'Pathogenic/Likely pathogenic', 'Benign', 'Likely benign',\n",
       "       'Conflicting interpretations of pathogenicity',\n",
       "       'Benign/Likely benign', 'drug response', 'other',\n",
       "       'Likely risk allele', 'risk factor',\n",
       "       'Pathogenic/Likely pathogenic/Established risk allele',\n",
       "       'Established risk allele', 'Pathogenic/Pathogenic, low penetrance',\n",
       "       'association'], dtype=object)"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data.clinical_sig.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Variation', 'single nucleotide variant', 'copy number gain'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data.variant_type.unique()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Process clinical significance"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "['Pathogenic', 'Uncertain significance', 'Likely pathogenic',\n",
    "       'Pathogenic/Likely pathogenic', 'Benign', 'Likely benign',\n",
    "       'Conflicting interpretations of pathogenicity',\n",
    "       'Benign/Likely benign', 'drug response', 'other',\n",
    "       'Likely risk allele', 'risk factor',\n",
    "       'Pathogenic/Likely pathogenic/Established risk allele',\n",
    "       'Established risk allele', 'Pathogenic/Pathogenic, low penetrance',\n",
    "       'association']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "def if_positive_or_negative(string_list):\n",
    "    label=[]\n",
    "    for string in string_list:\n",
    "        if string in ['Pathogenic', \n",
    "                    'Pathogenic/Likely pathogenic', \n",
    "                    'probable-pathogenic',\n",
    "                    'Likely pathogenic', \n",
    "                    'pathologic', \n",
    "                    'pathogenic',\n",
    "                    'likely pathogenic',\n",
    "                    'Pathogenic/Likely pathogenic/Established risk allele',\n",
    "                    'likely pathogenic - adrenal pheochromocytoma',\n",
    "                    'Pathogenic/Pathogenic, low penetrance',\n",
    "                    'Pathogenic, low penetrance']:\n",
    "            label.append(1)\n",
    "        elif string in ['Benign',\n",
    "                    'Likely benign',\n",
    "                    'Likely Benign',\n",
    "                    'Benign/Likely benign',\n",
    "                    'non-pathogenic', \n",
    "                    'benign', 'probable-non-pathogenic', 'Likely Benign', 'probably not pathogenic',\n",
    "        ]:\n",
    "            label.append(-1)\n",
    "\n",
    "        else: label.append(0)\n",
    "    return label"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "data['label']=if_positive_or_negative(data['clinical_sig'])"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Generate Positive Clinvar File and Negative Clinvar File"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "data[data['label']==1].to_csv('/home/grads/z/zshuying/Documents/shuying/ppi_mutation/data/clinvar/clinvar_pos.csv',sep=';')\n",
    "data[data['label']==-1].to_csv('/home/grads/z/zshuying/Documents/shuying/ppi_mutation/data/clinvar/clinvar_neg.csv',sep=';')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>clinvar_id</th>\n",
       "      <th>review_status</th>\n",
       "      <th>clinical_sig</th>\n",
       "      <th>uniprot_kb</th>\n",
       "      <th>variant_type</th>\n",
       "      <th>hgvs_p</th>\n",
       "      <th>missense</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>SCV000361195</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>Benign</td>\n",
       "      <td>P0C7Q2#VAR_044331</td>\n",
       "      <td>Variation</td>\n",
       "      <td>P0C7Q2:p.Ala69Ser</td>\n",
       "      <td>missense variant</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>SCV001325434</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>Likely benign</td>\n",
       "      <td>O60393#VAR_036636</td>\n",
       "      <td>Variation</td>\n",
       "      <td>O60393:p.Arg355His</td>\n",
       "      <td>missense variant</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>SCV000451797</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>Likely benign</td>\n",
       "      <td>Q9NRM0#VAR_065775</td>\n",
       "      <td>Variation</td>\n",
       "      <td>Q9NRM0:p.Arg198Cys</td>\n",
       "      <td>missense variant</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>SCV001136513</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>Benign</td>\n",
       "      <td>Q9NP91#VAR_052068</td>\n",
       "      <td>Variation</td>\n",
       "      <td>Q9NP91:p.Thr199Met</td>\n",
       "      <td>missense variant</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>SCV001138159</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>Benign</td>\n",
       "      <td>P15169#VAR_042415</td>\n",
       "      <td>Variation</td>\n",
       "      <td>P15169:p.Gly178Asp</td>\n",
       "      <td>missense variant</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32189</th>\n",
       "      <td>SCV002738545</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>Benign</td>\n",
       "      <td>P17813#VAR_014764</td>\n",
       "      <td>Variation</td>\n",
       "      <td>P17813:p.Asp366His</td>\n",
       "      <td>missense variant</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32191</th>\n",
       "      <td>SCV002612763</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>Benign</td>\n",
       "      <td>Q14896#VAR_074545</td>\n",
       "      <td>Variation</td>\n",
       "      <td>Q14896:p.Arg1138His</td>\n",
       "      <td>missense variant</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32192</th>\n",
       "      <td>SCV002613575</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>Likely benign</td>\n",
       "      <td>Q14896#VAR_020571</td>\n",
       "      <td>Variation</td>\n",
       "      <td>Q14896:p.Leu383Val</td>\n",
       "      <td>missense variant</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32196</th>\n",
       "      <td>SCV002614959</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>Benign</td>\n",
       "      <td>Q8TAB3#VAR_067489</td>\n",
       "      <td>Variation</td>\n",
       "      <td>Q8TAB3:p.Asn1134His</td>\n",
       "      <td>missense variant</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32200</th>\n",
       "      <td>SCV002618433</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>Likely benign</td>\n",
       "      <td>Q8WZ42#VAR_040089</td>\n",
       "      <td>Variation</td>\n",
       "      <td>Q8WZ42:p.Val1202Ala</td>\n",
       "      <td>missense variant</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>12563 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         clinvar_id                        review_status   clinical_sig  \\\n",
       "19     SCV000361195  criteria provided, single submitter         Benign   \n",
       "21     SCV001325434  criteria provided, single submitter  Likely benign   \n",
       "55     SCV000451797  criteria provided, single submitter  Likely benign   \n",
       "58     SCV001136513  criteria provided, single submitter         Benign   \n",
       "79     SCV001138159  criteria provided, single submitter         Benign   \n",
       "...             ...                                  ...            ...   \n",
       "32189  SCV002738545  criteria provided, single submitter         Benign   \n",
       "32191  SCV002612763  criteria provided, single submitter         Benign   \n",
       "32192  SCV002613575  criteria provided, single submitter  Likely benign   \n",
       "32196  SCV002614959  criteria provided, single submitter         Benign   \n",
       "32200  SCV002618433  criteria provided, single submitter  Likely benign   \n",
       "\n",
       "              uniprot_kb variant_type               hgvs_p          missense  \\\n",
       "19     P0C7Q2#VAR_044331    Variation    P0C7Q2:p.Ala69Ser  missense variant   \n",
       "21     O60393#VAR_036636    Variation   O60393:p.Arg355His  missense variant   \n",
       "55     Q9NRM0#VAR_065775    Variation   Q9NRM0:p.Arg198Cys  missense variant   \n",
       "58     Q9NP91#VAR_052068    Variation   Q9NP91:p.Thr199Met  missense variant   \n",
       "79     P15169#VAR_042415    Variation   P15169:p.Gly178Asp  missense variant   \n",
       "...                  ...          ...                  ...               ...   \n",
       "32189  P17813#VAR_014764    Variation   P17813:p.Asp366His  missense variant   \n",
       "32191  Q14896#VAR_074545    Variation  Q14896:p.Arg1138His  missense variant   \n",
       "32192  Q14896#VAR_020571    Variation   Q14896:p.Leu383Val  missense variant   \n",
       "32196  Q8TAB3#VAR_067489    Variation  Q8TAB3:p.Asn1134His  missense variant   \n",
       "32200  Q8WZ42#VAR_040089    Variation  Q8WZ42:p.Val1202Ala  missense variant   \n",
       "\n",
       "       label  \n",
       "19        -1  \n",
       "21        -1  \n",
       "55        -1  \n",
       "58        -1  \n",
       "79        -1  \n",
       "...      ...  \n",
       "32189     -1  \n",
       "32191     -1  \n",
       "32192     -1  \n",
       "32196     -1  \n",
       "32200     -1  \n",
       "\n",
       "[12563 rows x 8 columns]"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data[data['label']==-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def generate_example(start,end):\n",
    "    with open('/home/grads/z/zshuying/Documents/shuying/ppi_mutation/data/clinvar/ClinVarFullRelease_2021-06.xml', 'r') as f_in:\n",
    "        with open('/home/grads/z/zshuying/Documents/shuying/ppi_mutation/data/clinvar/example_'+str(start)+'_'+str(end)+'.xml','w') as f_out:\n",
    "            line_count=0\n",
    "            while True:\n",
    "                line=f_in.readline()\n",
    "                line_count+=1\n",
    "                if line_count>start and line_count<end:\n",
    "                    f_out.write(line)\n",
    "                if line_count>end:\n",
    "                    break\n",
    "\n",
    "def find_line(content):\n",
    "    with open('/home/grads/z/zshuying/Documents/shuying/ppi_mutation/data/clinvar/ClinVarFullRelease_2021-06.xml', 'r') as f_in:\n",
    "            line_count=0\n",
    "\n",
    "            while True:\n",
    "                line=f_in.readline()\n",
    "                line_count+=1\n",
    "\n",
    "                if content in line:\n",
    "                    print(line_count)\n",
    "                    break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "generate_example(45667191,45669191)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#AlleleID</th>\n",
       "      <th>Type</th>\n",
       "      <th>Name</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>GeneSymbol</th>\n",
       "      <th>HGNC_ID</th>\n",
       "      <th>ClinicalSignificance</th>\n",
       "      <th>ClinSigSimple</th>\n",
       "      <th>LastEvaluated</th>\n",
       "      <th>RS# (dbSNP)</th>\n",
       "      <th>...</th>\n",
       "      <th>ReviewStatus</th>\n",
       "      <th>NumberSubmitters</th>\n",
       "      <th>Guidelines</th>\n",
       "      <th>TestedInGTR</th>\n",
       "      <th>OtherIDs</th>\n",
       "      <th>SubmitterCategories</th>\n",
       "      <th>VariationID</th>\n",
       "      <th>PositionVCF</th>\n",
       "      <th>ReferenceAlleleVCF</th>\n",
       "      <th>AlternateAlleleVCF</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>15041</td>\n",
       "      <td>Indel</td>\n",
       "      <td>NM_014855.3(AP5Z1):c.80_83delinsTGCTGTAAACTGTA...</td>\n",
       "      <td>9907</td>\n",
       "      <td>AP5Z1</td>\n",
       "      <td>HGNC:22197</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>397704705</td>\n",
       "      <td>...</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>2</td>\n",
       "      <td>-</td>\n",
       "      <td>N</td>\n",
       "      <td>ClinGen:CA215070,OMIM:613653.0001</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>4820844</td>\n",
       "      <td>GGAT</td>\n",
       "      <td>TGCTGTAAACTGTAACTGTAAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>15041</td>\n",
       "      <td>Indel</td>\n",
       "      <td>NM_014855.3(AP5Z1):c.80_83delinsTGCTGTAAACTGTA...</td>\n",
       "      <td>9907</td>\n",
       "      <td>AP5Z1</td>\n",
       "      <td>HGNC:22197</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>397704705</td>\n",
       "      <td>...</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>2</td>\n",
       "      <td>-</td>\n",
       "      <td>N</td>\n",
       "      <td>ClinGen:CA215070,OMIM:613653.0001</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>4781213</td>\n",
       "      <td>GGAT</td>\n",
       "      <td>TGCTGTAAACTGTAACTGTAAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>15042</td>\n",
       "      <td>Deletion</td>\n",
       "      <td>NM_014855.3(AP5Z1):c.1413_1426del (p.Leu473fs)</td>\n",
       "      <td>9907</td>\n",
       "      <td>AP5Z1</td>\n",
       "      <td>HGNC:22197</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>Jun 29, 2010</td>\n",
       "      <td>397704709</td>\n",
       "      <td>...</td>\n",
       "      <td>no assertion criteria provided</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>N</td>\n",
       "      <td>ClinGen:CA215072,OMIM:613653.0002</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>4827360</td>\n",
       "      <td>GCTGCTGGACCTGCC</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>15042</td>\n",
       "      <td>Deletion</td>\n",
       "      <td>NM_014855.3(AP5Z1):c.1413_1426del (p.Leu473fs)</td>\n",
       "      <td>9907</td>\n",
       "      <td>AP5Z1</td>\n",
       "      <td>HGNC:22197</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>Jun 29, 2010</td>\n",
       "      <td>397704709</td>\n",
       "      <td>...</td>\n",
       "      <td>no assertion criteria provided</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>N</td>\n",
       "      <td>ClinGen:CA215072,OMIM:613653.0002</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>4787729</td>\n",
       "      <td>GCTGCTGGACCTGCC</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>15043</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_014630.3(ZNF592):c.3136G&gt;A (p.Gly1046Arg)</td>\n",
       "      <td>9640</td>\n",
       "      <td>ZNF592</td>\n",
       "      <td>HGNC:28986</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>0</td>\n",
       "      <td>Jun 29, 2015</td>\n",
       "      <td>150829393</td>\n",
       "      <td>...</td>\n",
       "      <td>no assertion criteria provided</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>N</td>\n",
       "      <td>OMIM:613624.0001,ClinGen:CA210674,UniProtKB:Q9...</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>85342440</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   #AlleleID                       Type  \\\n",
       "0      15041                      Indel   \n",
       "1      15041                      Indel   \n",
       "2      15042                   Deletion   \n",
       "3      15042                   Deletion   \n",
       "4      15043  single nucleotide variant   \n",
       "\n",
       "                                                Name  GeneID GeneSymbol  \\\n",
       "0  NM_014855.3(AP5Z1):c.80_83delinsTGCTGTAAACTGTA...    9907      AP5Z1   \n",
       "1  NM_014855.3(AP5Z1):c.80_83delinsTGCTGTAAACTGTA...    9907      AP5Z1   \n",
       "2     NM_014855.3(AP5Z1):c.1413_1426del (p.Leu473fs)    9907      AP5Z1   \n",
       "3     NM_014855.3(AP5Z1):c.1413_1426del (p.Leu473fs)    9907      AP5Z1   \n",
       "4       NM_014630.3(ZNF592):c.3136G>A (p.Gly1046Arg)    9640     ZNF592   \n",
       "\n",
       "      HGNC_ID    ClinicalSignificance  ClinSigSimple LastEvaluated  \\\n",
       "0  HGNC:22197              Pathogenic              1             -   \n",
       "1  HGNC:22197              Pathogenic              1             -   \n",
       "2  HGNC:22197              Pathogenic              1  Jun 29, 2010   \n",
       "3  HGNC:22197              Pathogenic              1  Jun 29, 2010   \n",
       "4  HGNC:28986  Uncertain significance              0  Jun 29, 2015   \n",
       "\n",
       "   RS# (dbSNP)  ...                         ReviewStatus NumberSubmitters  \\\n",
       "0    397704705  ...  criteria provided, single submitter                2   \n",
       "1    397704705  ...  criteria provided, single submitter                2   \n",
       "2    397704709  ...       no assertion criteria provided                1   \n",
       "3    397704709  ...       no assertion criteria provided                1   \n",
       "4    150829393  ...       no assertion criteria provided                1   \n",
       "\n",
       "  Guidelines TestedInGTR                                           OtherIDs  \\\n",
       "0          -           N                  ClinGen:CA215070,OMIM:613653.0001   \n",
       "1          -           N                  ClinGen:CA215070,OMIM:613653.0001   \n",
       "2          -           N                  ClinGen:CA215072,OMIM:613653.0002   \n",
       "3          -           N                  ClinGen:CA215072,OMIM:613653.0002   \n",
       "4          -           N  OMIM:613624.0001,ClinGen:CA210674,UniProtKB:Q9...   \n",
       "\n",
       "  SubmitterCategories VariationID PositionVCF ReferenceAlleleVCF  \\\n",
       "0                   3           2     4820844               GGAT   \n",
       "1                   3           2     4781213               GGAT   \n",
       "2                   1           3     4827360    GCTGCTGGACCTGCC   \n",
       "3                   1           3     4787729    GCTGCTGGACCTGCC   \n",
       "4                   1           4    85342440                  G   \n",
       "\n",
       "       AlternateAlleleVCF  \n",
       "0  TGCTGTAAACTGTAACTGTAAA  \n",
       "1  TGCTGTAAACTGTAACTGTAAA  \n",
       "2                       G  \n",
       "3                       G  \n",
       "4                       A  \n",
       "\n",
       "[5 rows x 34 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['#AlleleID', 'Type', 'Name', 'GeneID', 'GeneSymbol', 'HGNC_ID',\n",
       "       'ClinicalSignificance', 'ClinSigSimple', 'LastEvaluated', 'RS# (dbSNP)',\n",
       "       'nsv/esv (dbVar)', 'RCVaccession', 'PhenotypeIDS', 'PhenotypeList',\n",
       "       'Origin', 'OriginSimple', 'Assembly', 'ChromosomeAccession',\n",
       "       'Chromosome', 'Start', 'Stop', 'ReferenceAllele', 'AlternateAllele',\n",
       "       'Cytogenetic', 'ReviewStatus', 'NumberSubmitters', 'Guidelines',\n",
       "       'TestedInGTR', 'OtherIDs', 'SubmitterCategories', 'VariationID',\n",
       "       'PositionVCF', 'ReferenceAlleleVCF', 'AlternateAlleleVCF'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#AlleleID</th>\n",
       "      <th>Type</th>\n",
       "      <th>Name</th>\n",
       "      <th>ClinicalSignificance</th>\n",
       "      <th>ClinSigSimple</th>\n",
       "      <th>LastEvaluated</th>\n",
       "      <th>RCVaccession</th>\n",
       "      <th>PhenotypeIDS</th>\n",
       "      <th>PhenotypeList</th>\n",
       "      <th>Origin</th>\n",
       "      <th>OriginSimple</th>\n",
       "      <th>Assembly</th>\n",
       "      <th>ReviewStatus</th>\n",
       "      <th>NumberSubmitters</th>\n",
       "      <th>Guidelines</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>15041</td>\n",
       "      <td>Indel</td>\n",
       "      <td>NM_014855.3(AP5Z1):c.80_83delinsTGCTGTAAACTGTA...</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>RCV000000012</td>\n",
       "      <td>MONDO:MONDO:0013342,MedGen:C3150901,OMIM:61364...</td>\n",
       "      <td>Hereditary spastic paraplegia 48</td>\n",
       "      <td>germline;unknown</td>\n",
       "      <td>germline</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>2</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>15041</td>\n",
       "      <td>Indel</td>\n",
       "      <td>NM_014855.3(AP5Z1):c.80_83delinsTGCTGTAAACTGTA...</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>RCV000000012</td>\n",
       "      <td>MONDO:MONDO:0013342,MedGen:C3150901,OMIM:61364...</td>\n",
       "      <td>Hereditary spastic paraplegia 48</td>\n",
       "      <td>germline;unknown</td>\n",
       "      <td>germline</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>2</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>15042</td>\n",
       "      <td>Deletion</td>\n",
       "      <td>NM_014855.3(AP5Z1):c.1413_1426del (p.Leu473fs)</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>Jun 29, 2010</td>\n",
       "      <td>RCV000000013</td>\n",
       "      <td>MONDO:MONDO:0013342,MedGen:C3150901,OMIM:61364...</td>\n",
       "      <td>Hereditary spastic paraplegia 48</td>\n",
       "      <td>germline</td>\n",
       "      <td>germline</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>no assertion criteria provided</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>15042</td>\n",
       "      <td>Deletion</td>\n",
       "      <td>NM_014855.3(AP5Z1):c.1413_1426del (p.Leu473fs)</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>Jun 29, 2010</td>\n",
       "      <td>RCV000000013</td>\n",
       "      <td>MONDO:MONDO:0013342,MedGen:C3150901,OMIM:61364...</td>\n",
       "      <td>Hereditary spastic paraplegia 48</td>\n",
       "      <td>germline</td>\n",
       "      <td>germline</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>no assertion criteria provided</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>15043</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_014630.3(ZNF592):c.3136G&gt;A (p.Gly1046Arg)</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>0</td>\n",
       "      <td>Jun 29, 2015</td>\n",
       "      <td>RCV000000014</td>\n",
       "      <td>MONDO:MONDO:0033005,MedGen:C4551772,OMIM:25130...</td>\n",
       "      <td>Galloway-Mowat syndrome 1</td>\n",
       "      <td>germline</td>\n",
       "      <td>germline</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>no assertion criteria provided</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   #AlleleID                       Type  \\\n",
       "0      15041                      Indel   \n",
       "1      15041                      Indel   \n",
       "2      15042                   Deletion   \n",
       "3      15042                   Deletion   \n",
       "4      15043  single nucleotide variant   \n",
       "\n",
       "                                                Name    ClinicalSignificance  \\\n",
       "0  NM_014855.3(AP5Z1):c.80_83delinsTGCTGTAAACTGTA...              Pathogenic   \n",
       "1  NM_014855.3(AP5Z1):c.80_83delinsTGCTGTAAACTGTA...              Pathogenic   \n",
       "2     NM_014855.3(AP5Z1):c.1413_1426del (p.Leu473fs)              Pathogenic   \n",
       "3     NM_014855.3(AP5Z1):c.1413_1426del (p.Leu473fs)              Pathogenic   \n",
       "4       NM_014630.3(ZNF592):c.3136G>A (p.Gly1046Arg)  Uncertain significance   \n",
       "\n",
       "   ClinSigSimple LastEvaluated  RCVaccession  \\\n",
       "0              1             -  RCV000000012   \n",
       "1              1             -  RCV000000012   \n",
       "2              1  Jun 29, 2010  RCV000000013   \n",
       "3              1  Jun 29, 2010  RCV000000013   \n",
       "4              0  Jun 29, 2015  RCV000000014   \n",
       "\n",
       "                                        PhenotypeIDS  \\\n",
       "0  MONDO:MONDO:0013342,MedGen:C3150901,OMIM:61364...   \n",
       "1  MONDO:MONDO:0013342,MedGen:C3150901,OMIM:61364...   \n",
       "2  MONDO:MONDO:0013342,MedGen:C3150901,OMIM:61364...   \n",
       "3  MONDO:MONDO:0013342,MedGen:C3150901,OMIM:61364...   \n",
       "4  MONDO:MONDO:0033005,MedGen:C4551772,OMIM:25130...   \n",
       "\n",
       "                      PhenotypeList            Origin OriginSimple Assembly  \\\n",
       "0  Hereditary spastic paraplegia 48  germline;unknown     germline   GRCh37   \n",
       "1  Hereditary spastic paraplegia 48  germline;unknown     germline   GRCh38   \n",
       "2  Hereditary spastic paraplegia 48          germline     germline   GRCh37   \n",
       "3  Hereditary spastic paraplegia 48          germline     germline   GRCh38   \n",
       "4         Galloway-Mowat syndrome 1          germline     germline   GRCh37   \n",
       "\n",
       "                          ReviewStatus  NumberSubmitters Guidelines  \n",
       "0  criteria provided, single submitter                 2          -  \n",
       "1  criteria provided, single submitter                 2          -  \n",
       "2       no assertion criteria provided                 1          -  \n",
       "3       no assertion criteria provided                 1          -  \n",
       "4       no assertion criteria provided                 1          -  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f.loc[:,['#AlleleID', 'Type', 'Name','ClinicalSignificance', 'ClinSigSimple', 'LastEvaluated','RCVaccession', 'PhenotypeIDS', 'PhenotypeList','Origin', 'OriginSimple', 'Assembly','ReviewStatus', 'NumberSubmitters', 'Guidelines' ]].head()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Dealing with the variant summary file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_124161/2855674157.py:2: DtypeWarning: Columns (18) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  f=pd.read_csv('/home/grads/z/zshuying/Documents/shuying/ppi_mutation/data/clinvar/variant_summary.txt',sep='\\t')\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "f=pd.read_csv('/home/grads/z/zshuying/Documents/shuying/ppi_mutation/data/clinvar/variant_summary.txt',sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#AlleleID</th>\n",
       "      <th>Type</th>\n",
       "      <th>Name</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>GeneSymbol</th>\n",
       "      <th>HGNC_ID</th>\n",
       "      <th>ClinicalSignificance</th>\n",
       "      <th>ClinSigSimple</th>\n",
       "      <th>LastEvaluated</th>\n",
       "      <th>RS# (dbSNP)</th>\n",
       "      <th>...</th>\n",
       "      <th>ReviewStatus</th>\n",
       "      <th>NumberSubmitters</th>\n",
       "      <th>Guidelines</th>\n",
       "      <th>TestedInGTR</th>\n",
       "      <th>OtherIDs</th>\n",
       "      <th>SubmitterCategories</th>\n",
       "      <th>VariationID</th>\n",
       "      <th>PositionVCF</th>\n",
       "      <th>ReferenceAlleleVCF</th>\n",
       "      <th>AlternateAlleleVCF</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>15041</td>\n",
       "      <td>Indel</td>\n",
       "      <td>NM_014855.3(AP5Z1):c.80_83delinsTGCTGTAAACTGTA...</td>\n",
       "      <td>9907</td>\n",
       "      <td>AP5Z1</td>\n",
       "      <td>HGNC:22197</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>397704705</td>\n",
       "      <td>...</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>2</td>\n",
       "      <td>-</td>\n",
       "      <td>N</td>\n",
       "      <td>ClinGen:CA215070,OMIM:613653.0001</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>4820844</td>\n",
       "      <td>GGAT</td>\n",
       "      <td>TGCTGTAAACTGTAACTGTAAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>15041</td>\n",
       "      <td>Indel</td>\n",
       "      <td>NM_014855.3(AP5Z1):c.80_83delinsTGCTGTAAACTGTA...</td>\n",
       "      <td>9907</td>\n",
       "      <td>AP5Z1</td>\n",
       "      <td>HGNC:22197</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>397704705</td>\n",
       "      <td>...</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>2</td>\n",
       "      <td>-</td>\n",
       "      <td>N</td>\n",
       "      <td>ClinGen:CA215070,OMIM:613653.0001</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>4781213</td>\n",
       "      <td>GGAT</td>\n",
       "      <td>TGCTGTAAACTGTAACTGTAAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>15042</td>\n",
       "      <td>Deletion</td>\n",
       "      <td>NM_014855.3(AP5Z1):c.1413_1426del (p.Leu473fs)</td>\n",
       "      <td>9907</td>\n",
       "      <td>AP5Z1</td>\n",
       "      <td>HGNC:22197</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>Jun 29, 2010</td>\n",
       "      <td>397704709</td>\n",
       "      <td>...</td>\n",
       "      <td>no assertion criteria provided</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>N</td>\n",
       "      <td>ClinGen:CA215072,OMIM:613653.0002</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>4827360</td>\n",
       "      <td>GCTGCTGGACCTGCC</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>15042</td>\n",
       "      <td>Deletion</td>\n",
       "      <td>NM_014855.3(AP5Z1):c.1413_1426del (p.Leu473fs)</td>\n",
       "      <td>9907</td>\n",
       "      <td>AP5Z1</td>\n",
       "      <td>HGNC:22197</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>Jun 29, 2010</td>\n",
       "      <td>397704709</td>\n",
       "      <td>...</td>\n",
       "      <td>no assertion criteria provided</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>N</td>\n",
       "      <td>ClinGen:CA215072,OMIM:613653.0002</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>4787729</td>\n",
       "      <td>GCTGCTGGACCTGCC</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>15043</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_014630.3(ZNF592):c.3136G&gt;A (p.Gly1046Arg)</td>\n",
       "      <td>9640</td>\n",
       "      <td>ZNF592</td>\n",
       "      <td>HGNC:28986</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>0</td>\n",
       "      <td>Jun 29, 2015</td>\n",
       "      <td>150829393</td>\n",
       "      <td>...</td>\n",
       "      <td>no assertion criteria provided</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>N</td>\n",
       "      <td>OMIM:613624.0001,ClinGen:CA210674,UniProtKB:Q9...</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>85342440</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   #AlleleID                       Type  \\\n",
       "0      15041                      Indel   \n",
       "1      15041                      Indel   \n",
       "2      15042                   Deletion   \n",
       "3      15042                   Deletion   \n",
       "4      15043  single nucleotide variant   \n",
       "\n",
       "                                                Name  GeneID GeneSymbol  \\\n",
       "0  NM_014855.3(AP5Z1):c.80_83delinsTGCTGTAAACTGTA...    9907      AP5Z1   \n",
       "1  NM_014855.3(AP5Z1):c.80_83delinsTGCTGTAAACTGTA...    9907      AP5Z1   \n",
       "2     NM_014855.3(AP5Z1):c.1413_1426del (p.Leu473fs)    9907      AP5Z1   \n",
       "3     NM_014855.3(AP5Z1):c.1413_1426del (p.Leu473fs)    9907      AP5Z1   \n",
       "4       NM_014630.3(ZNF592):c.3136G>A (p.Gly1046Arg)    9640     ZNF592   \n",
       "\n",
       "      HGNC_ID    ClinicalSignificance  ClinSigSimple LastEvaluated  \\\n",
       "0  HGNC:22197              Pathogenic              1             -   \n",
       "1  HGNC:22197              Pathogenic              1             -   \n",
       "2  HGNC:22197              Pathogenic              1  Jun 29, 2010   \n",
       "3  HGNC:22197              Pathogenic              1  Jun 29, 2010   \n",
       "4  HGNC:28986  Uncertain significance              0  Jun 29, 2015   \n",
       "\n",
       "   RS# (dbSNP)  ...                         ReviewStatus NumberSubmitters  \\\n",
       "0    397704705  ...  criteria provided, single submitter                2   \n",
       "1    397704705  ...  criteria provided, single submitter                2   \n",
       "2    397704709  ...       no assertion criteria provided                1   \n",
       "3    397704709  ...       no assertion criteria provided                1   \n",
       "4    150829393  ...       no assertion criteria provided                1   \n",
       "\n",
       "  Guidelines TestedInGTR                                           OtherIDs  \\\n",
       "0          -           N                  ClinGen:CA215070,OMIM:613653.0001   \n",
       "1          -           N                  ClinGen:CA215070,OMIM:613653.0001   \n",
       "2          -           N                  ClinGen:CA215072,OMIM:613653.0002   \n",
       "3          -           N                  ClinGen:CA215072,OMIM:613653.0002   \n",
       "4          -           N  OMIM:613624.0001,ClinGen:CA210674,UniProtKB:Q9...   \n",
       "\n",
       "  SubmitterCategories VariationID PositionVCF ReferenceAlleleVCF  \\\n",
       "0                   3           2     4820844               GGAT   \n",
       "1                   3           2     4781213               GGAT   \n",
       "2                   1           3     4827360    GCTGCTGGACCTGCC   \n",
       "3                   1           3     4787729    GCTGCTGGACCTGCC   \n",
       "4                   1           4    85342440                  G   \n",
       "\n",
       "       AlternateAlleleVCF  \n",
       "0  TGCTGTAAACTGTAACTGTAAA  \n",
       "1  TGCTGTAAACTGTAACTGTAAA  \n",
       "2                       G  \n",
       "3                       G  \n",
       "4                       A  \n",
       "\n",
       "[5 rows x 34 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#AlleleID</th>\n",
       "      <th>Type</th>\n",
       "      <th>Name</th>\n",
       "      <th>label</th>\n",
       "      <th>ClinicalSignificance</th>\n",
       "      <th>ClinSigSimple</th>\n",
       "      <th>ReviewStatus</th>\n",
       "      <th>OtherIDs</th>\n",
       "      <th>LastEvaluated</th>\n",
       "      <th>RS# (dbSNP)</th>\n",
       "      <th>nsv/esv (dbVar)</th>\n",
       "      <th>RCVaccession</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>15044</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_017547.4(FOXRED1):c.694C&gt;T (p.Gln232Ter)</td>\n",
       "      <td>1</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>criteria provided, multiple submitters, no con...</td>\n",
       "      <td>ClinGen:CA113792,OMIM:613622.0001</td>\n",
       "      <td>Jun 28, 2022</td>\n",
       "      <td>267606829</td>\n",
       "      <td>-</td>\n",
       "      <td>RCV000000015|RCV000578659|RCV001194045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>15058</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_000410.4(HFE):c.848A&gt;C (p.Gln283Pro)</td>\n",
       "      <td>1</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>ClinGen:CA280949,UniProtKB:Q30201#VAR_037304,O...</td>\n",
       "      <td>Sep 12, 2022</td>\n",
       "      <td>111033563</td>\n",
       "      <td>-</td>\n",
       "      <td>RCV000000036|RCV001050090</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>15062</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_020779.4(WDR35):c.2590G&gt;A (p.Ala864Thr)</td>\n",
       "      <td>1</td>\n",
       "      <td>Likely pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>UniProtKB:Q9P2L0#VAR_064582,OMIM:613602.0004,C...</td>\n",
       "      <td>May 25, 2017</td>\n",
       "      <td>267607175</td>\n",
       "      <td>-</td>\n",
       "      <td>RCV000000040|RCV000508347</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>15066</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_001042472.3(ABHD12):c.1054C&gt;T (p.Arg352Ter)</td>\n",
       "      <td>1</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>criteria provided, multiple submitters, no con...</td>\n",
       "      <td>ClinGen:CA113811,OMIM:613599.0004</td>\n",
       "      <td>Jan 05, 2022</td>\n",
       "      <td>267606624</td>\n",
       "      <td>-</td>\n",
       "      <td>RCV000000044|RCV001208516</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>15069</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_138413.4(HOGA1):c.860G&gt;T (p.Gly287Val)</td>\n",
       "      <td>1</td>\n",
       "      <td>Pathogenic/Likely pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>criteria provided, multiple submitters, no con...</td>\n",
       "      <td>ClinGen:CA113813,UniProtKB:Q86XE5#VAR_064036,O...</td>\n",
       "      <td>Oct 13, 2022</td>\n",
       "      <td>138207257</td>\n",
       "      <td>-</td>\n",
       "      <td>RCV000000047|RCV000798240</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    #AlleleID                       Type  \\\n",
       "6       15044  single nucleotide variant   \n",
       "32      15058  single nucleotide variant   \n",
       "40      15062  single nucleotide variant   \n",
       "48      15066  single nucleotide variant   \n",
       "54      15069  single nucleotide variant   \n",
       "\n",
       "                                              Name  label  \\\n",
       "6      NM_017547.4(FOXRED1):c.694C>T (p.Gln232Ter)      1   \n",
       "32         NM_000410.4(HFE):c.848A>C (p.Gln283Pro)      1   \n",
       "40      NM_020779.4(WDR35):c.2590G>A (p.Ala864Thr)      1   \n",
       "48  NM_001042472.3(ABHD12):c.1054C>T (p.Arg352Ter)      1   \n",
       "54       NM_138413.4(HOGA1):c.860G>T (p.Gly287Val)      1   \n",
       "\n",
       "            ClinicalSignificance  ClinSigSimple  \\\n",
       "6                     Pathogenic              1   \n",
       "32                    Pathogenic              1   \n",
       "40             Likely pathogenic              1   \n",
       "48                    Pathogenic              1   \n",
       "54  Pathogenic/Likely pathogenic              1   \n",
       "\n",
       "                                         ReviewStatus  \\\n",
       "6   criteria provided, multiple submitters, no con...   \n",
       "32                criteria provided, single submitter   \n",
       "40                criteria provided, single submitter   \n",
       "48  criteria provided, multiple submitters, no con...   \n",
       "54  criteria provided, multiple submitters, no con...   \n",
       "\n",
       "                                             OtherIDs LastEvaluated  \\\n",
       "6                   ClinGen:CA113792,OMIM:613622.0001  Jun 28, 2022   \n",
       "32  ClinGen:CA280949,UniProtKB:Q30201#VAR_037304,O...  Sep 12, 2022   \n",
       "40  UniProtKB:Q9P2L0#VAR_064582,OMIM:613602.0004,C...  May 25, 2017   \n",
       "48                  ClinGen:CA113811,OMIM:613599.0004  Jan 05, 2022   \n",
       "54  ClinGen:CA113813,UniProtKB:Q86XE5#VAR_064036,O...  Oct 13, 2022   \n",
       "\n",
       "    RS# (dbSNP) nsv/esv (dbVar)                            RCVaccession  \n",
       "6     267606829               -  RCV000000015|RCV000578659|RCV001194045  \n",
       "32    111033563               -               RCV000000036|RCV001050090  \n",
       "40    267607175               -               RCV000000040|RCV000508347  \n",
       "48    267606624               -               RCV000000044|RCV001208516  \n",
       "54    138207257               -               RCV000000047|RCV000798240  "
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def if_positive_or_negative(string_list):\n",
    "    label=[]\n",
    "    for string in string_list:\n",
    "        if string in ['Pathogenic', \n",
    "                    'Pathogenic/Likely pathogenic', \n",
    "                    'probable-pathogenic',\n",
    "                    'Likely pathogenic', \n",
    "                    'pathologic', \n",
    "                    'pathogenic',\n",
    "                    'likely pathogenic',\n",
    "                    'Pathogenic/Likely pathogenic/Established risk allele',\n",
    "                    'likely pathogenic - adrenal pheochromocytoma',\n",
    "                    'Pathogenic/Pathogenic, low penetrance',\n",
    "                    'Pathogenic, low penetrance']:\n",
    "            label.append(1)\n",
    "        elif string in ['Benign',\n",
    "                    'Likely benign',\n",
    "                    'Likely Benign',\n",
    "                    'Benign/Likely benign',\n",
    "                    'non-pathogenic', \n",
    "                    'benign', 'probable-non-pathogenic', 'Likely Benign', 'probably not pathogenic',\n",
    "        ]:\n",
    "            label.append(-1)\n",
    "\n",
    "        else: label.append(0)\n",
    "    return label\n",
    "f['label']=if_positive_or_negative(f['ClinicalSignificance'])\n",
    "keep_cols=['#AlleleID', 'Type', 'Name','label', 'ClinicalSignificance', 'ClinSigSimple', 'ReviewStatus', 'OtherIDs', 'LastEvaluated', 'RS# (dbSNP)',\n",
    "       'nsv/esv (dbVar)', 'RCVaccession']\n",
    "keep_conditions=(f['Type']=='single nucleotide variant') &\\\n",
    "              (['no assertion' not in item for item in f['ReviewStatus'].tolist()]) &\\\n",
    "               (['p' in item for item in f['Name'].tolist()]) &\\\n",
    "               (f['label']!=0)\n",
    "f_simple=f[keep_conditions][keep_cols]\n",
    "f_simple.head()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  0%|          | 646/558616 [04:48<66:51:00,  2.32it/s] "
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Error found in https://rest.uniprot.org/uniprotkb/search?query=NM_001191016.3\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  0%|          | 1007/558616 [07:28<33:30:49,  4.62it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Error Finding Refseq\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  0%|          | 1032/558616 [07:38<98:10:26,  1.58it/s]"
     ]
    }
   ],
   "source": [
    "from tqdm import tqdm\n",
    "import multiprocessing\n",
    "num_processes = multiprocessing.cpu_count()\n",
    "chunk_size = int(f_simple.shape[0]/num_processes)\n",
    "global pbar\n",
    "pbar = tqdm(total=len(f_simple))\n",
    "f_simple['UniPort']=[0]*len(f_simple)\n",
    "chunks = [f_simple.iloc[f_simple.index[i:i + chunk_size]] for i in range(0, f_simple.shape[0], chunk_size)]\n",
    "def func(df):\n",
    "    for idx in df.index:\n",
    "        if 'UniProt' not in df.loc[idx,'OtherIDs']:\n",
    "                df.loc[idx,'UniPort']=get_uniprot_from_name(f_simple.loc[idx,'Name'])\n",
    "        pbar.update(1)\n",
    "    return df\n",
    "pool = multiprocessing.Pool(processes=num_processes)\n",
    "result = pool.map(func, chunks)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Int64Index([      6,      32,      40,      48,      54,      60,      62,\n",
       "                 64,      66,      70,\n",
       "            ...\n",
       "            4323230, 4323238, 4323259, 4323261, 4323285, 4323293, 4323305,\n",
       "            4323310, 4323312, 4323316],\n",
       "           dtype='int64', length=558616)"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f_simple.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Error found in https://rest.uniprot.org/uniprotkb/search?query=NM_001142446.2\n",
      "Error found in https://rest.uniprot.org/uniprotkb/search?query=NM_001142446.2\n",
      "Error found in https://rest.uniprot.org/uniprotkb/search?query=NM_001127255.2\n",
      "Error found in https://rest.uniprot.org/uniprotkb/search?query=NM_001127255.2\n",
      "Error found in https://rest.uniprot.org/uniprotkb/search?query=NM_001127255.2\n",
      "Error found in https://rest.uniprot.org/uniprotkb/search?query=NM_001127255.2\n",
      "Error found in https://rest.uniprot.org/uniprotkb/search?query=NM_001277059.2\n",
      "Error found in https://rest.uniprot.org/uniprotkb/search?query=NM_001277059.2\n",
      "Error found in https://rest.uniprot.org/uniprotkb/search?query=NM_001191016.3\n",
      "Error found in https://rest.uniprot.org/uniprotkb/search?query=NM_001191016.3\n",
      "Error found in https://rest.uniprot.org/uniprotkb/search?query=NM_199451.3\n",
      "Error found in https://rest.uniprot.org/uniprotkb/search?query=NM_199451.3\n",
      "Error found in https://rest.uniprot.org/uniprotkb/search?query=NM_001368067.1\n",
      "Error found in https://rest.uniprot.org/uniprotkb/search?query=NM_001368067.1\n",
      "Error found in https://rest.uniprot.org/uniprotkb/search?query=NM_001368067.1\n",
      "Error found in https://rest.uniprot.org/uniprotkb/search?query=NM_001368067.1\n",
      "Error found in https://rest.uniprot.org/uniprotkb/search?query=NM_001368067.1\n",
      "Error found in https://rest.uniprot.org/uniprotkb/search?query=NM_001368067.1\n",
      "Error found in https://rest.uniprot.org/uniprotkb/search?query=NM_001368067.1\n",
      "Error found in https://rest.uniprot.org/uniprotkb/search?query=NM_001368067.1\n",
      "Error found in https://rest.uniprot.org/uniprotkb/search?query=NM_001368067.1\n",
      "Error found in https://rest.uniprot.org/uniprotkb/search?query=NM_001368067.1\n",
      "Error found in https://rest.uniprot.org/uniprotkb/search?query=NM_213655.5\n",
      "Error found in https://rest.uniprot.org/uniprotkb/search?query=NM_213655.5\n",
      "Error found in https://rest.uniprot.org/uniprotkb/search?query=NM_213655.5\n",
      "Error found in https://rest.uniprot.org/uniprotkb/search?query=NM_213655.5\n",
      "Error found in https://rest.uniprot.org/uniprotkb/search?query=NM_176806.4\n",
      "Error found in https://rest.uniprot.org/uniprotkb/search?query=NM_176806.4\n",
      "Error found in https://rest.uniprot.org/uniprotkb/search?query=NM_001164277.2\n",
      "Error found in https://rest.uniprot.org/uniprotkb/search?query=NM_001164277.2\n",
      "Error found in https://rest.uniprot.org/uniprotkb/search?query=NM_001164277.2\n",
      "Error found in https://rest.uniprot.org/uniprotkb/search?query=NM_001164277.2\n",
      "Error found in https://rest.uniprot.org/uniprotkb/search?query=NM_001164277.2\n",
      "Error found in https://rest.uniprot.org/uniprotkb/search?query=NM_001164277.2\n",
      "Error found in https://rest.uniprot.org/uniprotkb/search?query=NM_001164277.2\n",
      "Error found in https://rest.uniprot.org/uniprotkb/search?query=NM_001164277.2\n",
      "Error found in https://rest.uniprot.org/uniprotkb/search?query=NM_001164277.2\n",
      "Error found in https://rest.uniprot.org/uniprotkb/search?query=NM_001164277.2\n",
      "Error found in https://rest.uniprot.org/uniprotkb/search?query=NM_001164277.2\n",
      "Error found in https://rest.uniprot.org/uniprotkb/search?query=NM_001164277.2\n",
      "Error found in https://rest.uniprot.org/uniprotkb/search?query=NM_001164277.2\n",
      "Error found in https://rest.uniprot.org/uniprotkb/search?query=NM_001164277.2\n",
      "Error found in https://rest.uniprot.org/uniprotkb/search?query=NM_001164277.2\n",
      "Error found in https://rest.uniprot.org/uniprotkb/search?query=NM_001164277.2\n",
      "Error found in https://rest.uniprot.org/uniprotkb/search?query=NR_001566.1\n",
      "Error found in https://rest.uniprot.org/uniprotkb/search?query=NR_001566.1\n",
      "Error found in https://rest.uniprot.org/uniprotkb/search?query=NR_001566.1\n",
      "Error found in https://rest.uniprot.org/uniprotkb/search?query=NR_001566.1\n",
      "Error Finding Refseq\n"
     ]
    },
    {
     "ename": "UnboundLocalError",
     "evalue": "local variable 'refseq' referenced before assignment",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mUnboundLocalError\u001b[0m                         Traceback (most recent call last)",
      "\u001b[1;32m/home/grads/z/zshuying/Documents/shuying/ppi_mutation/scripts/clinvar.ipynb Cell 28\u001b[0m in \u001b[0;36m<cell line: 1>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> <a href='vscode-notebook-cell://ssh-remote%2B10.125.185.12/home/grads/z/zshuying/Documents/shuying/ppi_mutation/scripts/clinvar.ipynb#X43sdnNjb2RlLXJlbW90ZQ%3D%3D?line=0'>1</a>\u001b[0m f_simple[\u001b[39m'\u001b[39m\u001b[39mUniProtID\u001b[39m\u001b[39m'\u001b[39m]\u001b[39m=\u001b[39m[get_uniprot_from_name(name) \u001b[39mfor\u001b[39;00m name \u001b[39min\u001b[39;00m f_simple[\u001b[39m'\u001b[39m\u001b[39mName\u001b[39m\u001b[39m'\u001b[39m]]\n",
      "\u001b[1;32m/home/grads/z/zshuying/Documents/shuying/ppi_mutation/scripts/clinvar.ipynb Cell 28\u001b[0m in \u001b[0;36m<listcomp>\u001b[0;34m(.0)\u001b[0m\n\u001b[0;32m----> <a href='vscode-notebook-cell://ssh-remote%2B10.125.185.12/home/grads/z/zshuying/Documents/shuying/ppi_mutation/scripts/clinvar.ipynb#X43sdnNjb2RlLXJlbW90ZQ%3D%3D?line=0'>1</a>\u001b[0m f_simple[\u001b[39m'\u001b[39m\u001b[39mUniProtID\u001b[39m\u001b[39m'\u001b[39m]\u001b[39m=\u001b[39m[get_uniprot_from_name(name) \u001b[39mfor\u001b[39;00m name \u001b[39min\u001b[39;00m f_simple[\u001b[39m'\u001b[39m\u001b[39mName\u001b[39m\u001b[39m'\u001b[39m]]\n",
      "\u001b[1;32m/home/grads/z/zshuying/Documents/shuying/ppi_mutation/scripts/clinvar.ipynb Cell 28\u001b[0m in \u001b[0;36mget_uniprot_from_name\u001b[0;34m(name)\u001b[0m\n\u001b[1;32m     <a href='vscode-notebook-cell://ssh-remote%2B10.125.185.12/home/grads/z/zshuying/Documents/shuying/ppi_mutation/scripts/clinvar.ipynb#X43sdnNjb2RlLXJlbW90ZQ%3D%3D?line=16'>17</a>\u001b[0m \u001b[39mexcept\u001b[39;00m:\n\u001b[1;32m     <a href='vscode-notebook-cell://ssh-remote%2B10.125.185.12/home/grads/z/zshuying/Documents/shuying/ppi_mutation/scripts/clinvar.ipynb#X43sdnNjb2RlLXJlbW90ZQ%3D%3D?line=17'>18</a>\u001b[0m     \u001b[39mprint\u001b[39m(\u001b[39m'\u001b[39m\u001b[39mError Finding Refseq\u001b[39m\u001b[39m'\u001b[39m)\n\u001b[0;32m---> <a href='vscode-notebook-cell://ssh-remote%2B10.125.185.12/home/grads/z/zshuying/Documents/shuying/ppi_mutation/scripts/clinvar.ipynb#X43sdnNjb2RlLXJlbW90ZQ%3D%3D?line=18'>19</a>\u001b[0m uniprot\u001b[39m=\u001b[39mfind_fasta_refseq(refseq)\n\u001b[1;32m     <a href='vscode-notebook-cell://ssh-remote%2B10.125.185.12/home/grads/z/zshuying/Documents/shuying/ppi_mutation/scripts/clinvar.ipynb#X43sdnNjb2RlLXJlbW90ZQ%3D%3D?line=19'>20</a>\u001b[0m \u001b[39mreturn\u001b[39;00m uniprot\n",
      "\u001b[0;31mUnboundLocalError\u001b[0m: local variable 'refseq' referenced before assignment"
     ]
    }
   ],
   "source": [
    "f_simple['UniProtID']=[get_uniprot_from_name(name) for name in f_simple['Name']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [],
   "source": [
    "f_simple=f_simple.drop_duplicates(subset=['#AlleleID'])\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Map RefSeqID to Uniprot ID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "import json\n",
    "import re\n",
    "def find_fasta_refseq(refseqID):\n",
    "    url='https://rest.uniprot.org/uniprotkb/search?query=%s'%refseqID\n",
    "    cnt=json.loads(requests.get(url).text)\n",
    "    try:\n",
    "        assert re.search('uniprot',cnt.get(\"results\")[0].get(\"entryType\"),re.IGNORECASE)\n",
    "        return cnt[\"results\"][0]['primaryAccession']\n",
    "    except:\n",
    "        print('Error found in '+url)\n",
    "        return 'Error found in '+url\n",
    "\n",
    "def get_uniprot_from_name(name):\n",
    "    try:\n",
    "        refseq=re.match(r'\\S*?(?=[\\(:])',name).group(0)\n",
    "        uniprot=find_fasta_refseq(refseq)\n",
    "\n",
    "    except:\n",
    "        print('Error Finding Refseq')\n",
    "        uniprot=find_fasta_refseq(refseq)\n",
    "    return uniprot "
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Extract UniProt from OtherIDs    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "def extract_uniprot(string):\n",
    "    try: m=re.match(r'.*UniProtKB:([0-9A-Z].*)#.*',string).group(1)\n",
    "    except AttributeError: \n",
    "        print(string)\n",
    "        m='0'\n",
    "    return m"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Merge multiprocess files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import re\n",
    "def merge(dir,prefix,out_name,compare=True):\n",
    "    d={}\n",
    "    import os\n",
    "    f_list=os.listdir(dir)\n",
    "    for f in f_list:\n",
    "        print(f)\n",
    "        proc,num=re.match(r'%s([0-9a-zA-Z]*)_([0-9a-zA-Z]*)_finished.csv'%prefix,f).group(1),\\\n",
    "        re.match(r'%s([0-9a-zA-Z]*)_([0-9a-zA-Z]*)_finished.csv'%prefix,f).group(2)\n",
    "        if compare:\n",
    "            if  d.get(proc) is None or int(num)> d.get(proc):\n",
    "                d[proc]=int(num)\n",
    "        else:\n",
    "            d[proc]=num\n",
    "    f_out_list=[]\n",
    "    for key in d.keys():\n",
    "        f_out_list.append(os.path.join(dir,'%s%s_%s_finished.csv'%(prefix,key,d[key])))\n",
    "    print(f_out_list)\n",
    "    dfs=[]\n",
    "    for f in f_out_list:\n",
    "        dfs.append(pd.read_csv(f,index_col='#AlleleID')\n",
    "                )\n",
    "    df_out=pd.concat(dfs)\n",
    "    df_out.to_csv(out_name)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2022-2-round1_proc_101605_5128_finished.csv\n",
      "2022-2-round1_proc_1086723_5128_finished.csv\n",
      "2022-2-round1_proc_1134089_5128_finished.csv\n",
      "2022-2-round1_proc_1162009_5128_finished.csv\n",
      "2022-2-round1_proc_1202281_5128_finished.csv\n",
      "2022-2-round1_proc_1322629_5128_finished.csv\n",
      "2022-2-round1_proc_1543645_5128_finished.csv\n",
      "2022-2-round1_proc_1670622_5128_finished.csv\n",
      "2022-2-round1_proc_183957_5128_finished.csv\n",
      "2022-2-round1_proc_1880543_5128_finished.csv\n",
      "2022-2-round1_proc_2083431_5000_finished.csv\n",
      "2022-2-round1_proc_2083431_5128_finished.csv\n",
      "2022-2-round1_proc_2332108_13_finished.csv\n",
      "2022-2-round1_proc_253695_5128_finished.csv\n",
      "2022-2-round1_proc_327311_5128_finished.csv\n",
      "2022-2-round1_proc_39665_5128_finished.csv\n",
      "2022-2-round1_proc_500469_5128_finished.csv\n",
      "2022-2-round1_proc_599586_5128_finished.csv\n",
      "2022-2-round1_proc_6_5128_finished.csv\n",
      "2022-2-round1_proc_671051_5128_finished.csv\n",
      "2022-2-round1_proc_792358_5128_finished.csv\n",
      "2022-2-round1_proc_923200_5128_finished.csv\n",
      "['/home/grads/z/zshuying/Documents/shuying/ppi_mutation/data/clinvar/2022-2/2022-2-round1_proc_101605_5128_finished.csv', '/home/grads/z/zshuying/Documents/shuying/ppi_mutation/data/clinvar/2022-2/2022-2-round1_proc_1086723_5128_finished.csv', '/home/grads/z/zshuying/Documents/shuying/ppi_mutation/data/clinvar/2022-2/2022-2-round1_proc_1134089_5128_finished.csv', '/home/grads/z/zshuying/Documents/shuying/ppi_mutation/data/clinvar/2022-2/2022-2-round1_proc_1162009_5128_finished.csv', '/home/grads/z/zshuying/Documents/shuying/ppi_mutation/data/clinvar/2022-2/2022-2-round1_proc_1202281_5128_finished.csv', '/home/grads/z/zshuying/Documents/shuying/ppi_mutation/data/clinvar/2022-2/2022-2-round1_proc_1322629_5128_finished.csv', '/home/grads/z/zshuying/Documents/shuying/ppi_mutation/data/clinvar/2022-2/2022-2-round1_proc_1543645_5128_finished.csv', '/home/grads/z/zshuying/Documents/shuying/ppi_mutation/data/clinvar/2022-2/2022-2-round1_proc_1670622_5128_finished.csv', '/home/grads/z/zshuying/Documents/shuying/ppi_mutation/data/clinvar/2022-2/2022-2-round1_proc_183957_5128_finished.csv', '/home/grads/z/zshuying/Documents/shuying/ppi_mutation/data/clinvar/2022-2/2022-2-round1_proc_1880543_5128_finished.csv', '/home/grads/z/zshuying/Documents/shuying/ppi_mutation/data/clinvar/2022-2/2022-2-round1_proc_2083431_5128_finished.csv', '/home/grads/z/zshuying/Documents/shuying/ppi_mutation/data/clinvar/2022-2/2022-2-round1_proc_2332108_13_finished.csv', '/home/grads/z/zshuying/Documents/shuying/ppi_mutation/data/clinvar/2022-2/2022-2-round1_proc_253695_5128_finished.csv', '/home/grads/z/zshuying/Documents/shuying/ppi_mutation/data/clinvar/2022-2/2022-2-round1_proc_327311_5128_finished.csv', '/home/grads/z/zshuying/Documents/shuying/ppi_mutation/data/clinvar/2022-2/2022-2-round1_proc_39665_5128_finished.csv', '/home/grads/z/zshuying/Documents/shuying/ppi_mutation/data/clinvar/2022-2/2022-2-round1_proc_500469_5128_finished.csv', '/home/grads/z/zshuying/Documents/shuying/ppi_mutation/data/clinvar/2022-2/2022-2-round1_proc_599586_5128_finished.csv', '/home/grads/z/zshuying/Documents/shuying/ppi_mutation/data/clinvar/2022-2/2022-2-round1_proc_6_5128_finished.csv', '/home/grads/z/zshuying/Documents/shuying/ppi_mutation/data/clinvar/2022-2/2022-2-round1_proc_671051_5128_finished.csv', '/home/grads/z/zshuying/Documents/shuying/ppi_mutation/data/clinvar/2022-2/2022-2-round1_proc_792358_5128_finished.csv', '/home/grads/z/zshuying/Documents/shuying/ppi_mutation/data/clinvar/2022-2/2022-2-round1_proc_923200_5128_finished.csv']\n"
     ]
    }
   ],
   "source": [
    "merge('/home/grads/z/zshuying/Documents/shuying/ppi_mutation/data/clinvar/2022-2','2022-2-round1_proc_','merged_2022_2.csv',False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "def exlude_synonymous(name):\n",
    "    #\"NM_017813.5(BPNT2):c.228C>A (p.Arg76=)\"\n",
    "    return bool(re.match(r'.*(\\(p\\.[a-zA-Z0-9]*=\\)).*',name))\n",
    "\n",
    "f2023=pd.read_csv('/home/grads/z/zshuying/Documents/shuying/ppi_mutation/scripts/merged_2023_2_26.csv',index_col='#AlleleID')\n",
    "f2023=f2023[[not exlude_synonymous(name) for name in f2023.Name.tolist()]]\n",
    "f2023.to_csv('/home/grads/z/zshuying/Documents/shuying/ppi_mutation/scripts/merged_2023_2_26.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "f2023=pd.read_csv('/home/grads/z/zshuying/Documents/shuying/ppi_mutation/scripts/merged_2023_2_26.csv',index_col='#AlleleID')\n",
    "f2022=pd.read_csv('/home/grads/z/zshuying/Documents/shuying/ppi_mutation/scripts/merged_2022_2.csv',index_col='#AlleleID')\n",
    "f2023_no_error=f2023[f2023['UniProt'].apply(lambda x: 'Error' not in x)]\n",
    "f2022_no_error=f2022[f2022['UniProt'].apply(lambda x: 'Error' not in x)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-1    146630\n",
      " 1     27375\n",
      "Name: label, dtype: int64\n",
      "-1    10686\n",
      " 1     8462\n",
      "Name: label, dtype: int64\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_108981/2630773320.py:4: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  print(f2023_no_error.loc[diff_1star]['label'].value_counts())\n",
      "/tmp/ipykernel_108981/2630773320.py:5: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  print(f2023_2star_no_error.loc[diff_2star]['label'].value_counts())\n"
     ]
    }
   ],
   "source": [
    "# len(f2022.index.unique()),len(f2023.index.unique()),len(set(f2023.index)-set(f2022.index))\n",
    "diff_1star=set(f2023_no_error.index)-set(f2022_no_error.index)\n",
    "diff_2star=set(f2023_2star_no_error.index)-set(f2022_2star_no_error.index)\n",
    "print(f2023_no_error.loc[diff_1star]['label'].value_counts())\n",
    "print(f2023_2star_no_error.loc[diff_2star]['label'].value_counts())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(33148, 51297, 19148)"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f2022_2star_no_error=f2022_no_error[['single' not in item for item in f2022_no_error['ReviewStatus'].tolist()]]\n",
    "f2023_2star_no_error=f2023_no_error[['single' not in item for item in f2023_no_error['ReviewStatus'].tolist()]]\n",
    "len(f2022_2star_no_error.index.unique()),len(f2023_2star_no_error.index.unique()),len(set(f2023_2star_no_error.index)-set(f2022_2star_no_error.index))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "f2022_2star_no_error.to_csv('clinvar_2starabove_2022_02.csv')\n",
    "f2023_2star_no_error.to_csv('clinvar_2starabove_2023_03.csv')\n",
    "f2023_no_error.to_csv('clinvar_1starabove_2023_02.csv')\n",
    "f2022_no_error.to_csv('clinvar_1starabove_2022_02.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "l=[\n",
    "    '/home/grads/z/zshuying/Documents/shuying/ppi_mutation/data/clinvar/2023_02_26_round4/2023-2-round4_proc_1089410_11400_finished.csv',\n",
    "    '/home/grads/z/zshuying/Documents/shuying/ppi_mutation/data/clinvar/2023_02_26_round4/2023-2-round4_proc_already_done_finished.csv',\n",
    "    '/home/grads/z/zshuying/Documents/shuying/ppi_mutation/data/clinvar/2023_02_26_round4/2023-2-round5_proc_134664_96_finished.csv',\n",
    "    '/home/grads/z/zshuying/Documents/shuying/ppi_mutation/data/clinvar/2023_02_26_round4/2023-2-round5_proc_already_done_finished.csv'\n",
    "]\n",
    "dfs=[]\n",
    "for f in l:\n",
    "    dfs.append(pd.read_csv(f,index_col='#AlleleID'))\n",
    "df=pd.concat(dfs,sort=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.drop_duplicates(subset='#AlleleID',)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ClinGen:CA284849,UniProtKB/Swiss-Prot:VAR_003185,OMIM:602671.0001\n",
      "ClinGen:CA219309,UniProtKB/Swiss-Prot:VAR_007850,OMIM:602671.0003\n",
      "ClinGen:CA219358,UniProtKB/Swiss-Prot:VAR_016840,OMIM:602671.0013\n",
      "ClinGen:CA219275,UniProtKB/Swiss-Prot:VAR_025601,OMIM:602671.0015\n",
      "ClinGen:CA254415,UniProtKB/Swiss-Prot:VAR_001491,OMIM:601011.0001\n",
      "ClinGen:CA119650,UniProtKB/Swiss-Prot:VAR_001492,OMIM:601011.0002\n",
      "ClinGen:CA254421,UniProtKB/Swiss-Prot:VAR_001494,OMIM:601011.0004\n",
      "ClinGen:CA254430,UniProtKB/Swiss-Prot:VAR_043837,OMIM:601011.0011\n",
      "ClinGen:CA254437,UniProtKB/Swiss-Prot:VAR_043825,OMIM:601011.0009\n",
      "ClinGen:CA254443,UniProtKB/Swiss-Prot:VAR_043835,OMIM:601011.0012\n",
      "ClinGen:CA254473,UniProtKB/Swiss-Prot:VAR_043821,OMIM:601011.0017\n",
      "ClinGen:CA119653,UniProtKB/Swiss-Prot:VAR_043826,OMIM:601011.0018\n",
      "ClinGen:CA254496,UniProtKB/Swiss-Prot:VAR_043830,OMIM:601011.0027\n",
      "ClinGen:CA256584,UniProtKB/Swiss-Prot:VAR_010111,OMIM:182389.0001\n",
      "ClinGen:CA256587,UniProtKB/Swiss-Prot:VAR_010110,OMIM:182389.0002\n",
      "ClinGen:CA256596,UniProtKB/Swiss-Prot:VAR_014273,OMIM:182389.0005\n",
      "ClinGen:CA256599,UniProtKB/Swiss-Prot:VAR_014270,OMIM:182389.0006\n",
      "ClinGen:CA281911,UniProtKB/Swiss-Prot:VAR_014268,OMIM:182389.0009\n",
      "ClinGen:CA256602,UniProtKB/Swiss-Prot:VAR_014271,OMIM:182389.0010\n",
      "ClinGen:CA256611,UniProtKB/Swiss-Prot:VAR_029717,OMIM:182389.0013\n",
      "ClinGen:CA256620,UniProtKB/Swiss-Prot:VAR_057997,OMIM:182389.0021\n",
      "ClinGen:CA129599,UniProtKB/Swiss-Prot:VAR_065236,OMIM:613113.0045\n",
      "ClinGen:CA129841,UniProtKB/Swiss-Prot:VAR_068224,OMIM:607211.0001\n",
      "ClinGen:CA218986,UniProtKB/Swiss-Prot:VAR_007284\n",
      "ClinGen:CA218994,UniProtKB/Swiss-Prot:VAR_054208\n",
      "ClinGen:CA218996,UniProtKB/Swiss-Prot:VAR_054209\n",
      "ClinGen:CA219014,UniProtKB/Swiss-Prot:VAR_054171\n",
      "ClinGen:CA219018,UniProtKB/Swiss-Prot:VAR_054172\n",
      "ClinGen:CA219022,UniProtKB/Swiss-Prot:VAR_054174\n",
      "ClinGen:CA219024,UniProtKB/Swiss-Prot:VAR_054175\n",
      "UniProtKB/Swiss-Prot:VAR_007255,ClinGen:CA219026\n",
      "ClinGen:CA219044,UniProtKB/Swiss-Prot:VAR_007263\n",
      "ClinGen:CA219046,UniProtKB/Swiss-Prot:VAR_054182\n",
      "ClinGen:CA219052,UniProtKB/Swiss-Prot:VAR_007268\n",
      "ClinGen:CA219054,UniProtKB/Swiss-Prot:VAR_054186\n",
      "ClinGen:CA219056,UniProtKB/Swiss-Prot:VAR_007270\n",
      "ClinGen:CA219060,UniProtKB/Swiss-Prot:VAR_054190\n",
      "ClinGen:CA219062,UniProtKB/Swiss-Prot:VAR_054192\n",
      "ClinGen:CA219064,UniProtKB/Swiss-Prot:VAR_054194\n",
      "ClinGen:CA219070,UniProtKB/Swiss-Prot:VAR_054196\n",
      "ClinGen:CA219080,UniProtKB/Swiss-Prot:VAR_067417\n",
      "ClinGen:CA219082,UniProtKB/Swiss-Prot:VAR_054199\n",
      "ClinGen:CA219084,UniProtKB/Swiss-Prot:VAR_054200\n",
      "ClinGen:CA219086,UniProtKB/Swiss-Prot:VAR_007275\n",
      "ClinGen:CA219188,UniProtKB/Swiss-Prot:VAR_065662\n",
      "ClinGen:CA219267,UniProtKB/Swiss-Prot:VAR_068835\n",
      "ClinGen:CA219300,UniProtKB/Swiss-Prot:VAR_025588\n",
      "ClinGen:CA219316,UniProtKB/Swiss-Prot:VAR_066395\n",
      "ClinGen:CA219319,UniProtKB/Swiss-Prot:VAR_003184\n",
      "ClinGen:CA219322,UniProtKB/Swiss-Prot:VAR_025592\n",
      "ClinGen:CA219325,UniProtKB/Swiss-Prot:VAR_025593\n",
      "ClinGen:CA219331,UniProtKB/Swiss-Prot:VAR_025595\n",
      "ClinGen:CA219334,UniProtKB/Swiss-Prot:VAR_032113\n",
      "ClinGen:CA219337,UniProtKB/Swiss-Prot:VAR_025581\n",
      "ClinGen:CA219343,UniProtKB/Swiss-Prot:VAR_025582\n",
      "ClinGen:CA219352,UniProtKB/Swiss-Prot:VAR_025584\n",
      "ClinGen:CA219364,UniProtKB/Swiss-Prot:VAR_025599\n",
      "ClinGen:CA152091,UniProtKB/Swiss-Prot:VAR_014459\n",
      "ClinGen:CA152094,UniProtKB/Swiss-Prot:VAR_043828\n",
      "ClinGen:CA147061,UniProtKB/Swiss-Prot:VAR_014462\n",
      "ClinGen:CA266053,UniProtKB/Swiss-Prot:VAR_043838\n",
      "ClinGen:CA266071,UniProtKB/Swiss-Prot:VAR_043820\n",
      "ClinGen:CA266080,UniProtKB/Swiss-Prot:VAR_043822\n",
      "ClinGen:CA284895,UniProtKB/Swiss-Prot:VAR_064304\n",
      "ClinGen:CA284910,UniProtKB/Swiss-Prot:VAR_029683\n",
      "ClinGen:CA284913,UniProtKB/Swiss-Prot:VAR_064251\n",
      "ClinGen:CA284925,UniProtKB/Swiss-Prot:VAR_043360\n",
      "ClinGen:CA284928,UniProtKB/Swiss-Prot:VAR_029693\n",
      "ClinGen:CA284931,UniProtKB/Swiss-Prot:VAR_064311\n",
      "ClinGen:CA284934,UniProtKB/Swiss-Prot:VAR_043362\n",
      "ClinGen:CA284940,UniProtKB/Swiss-Prot:VAR_029699\n",
      "ClinGen:CA284943,UniProtKB/Swiss-Prot:VAR_064260\n",
      "ClinGen:CA284949,UniProtKB/Swiss-Prot:VAR_064313\n",
      "ClinGen:CA284958,UniProtKB/Swiss-Prot:VAR_064348\n",
      "ClinGen:CA284961,UniProtKB/Swiss-Prot:VAR_064315\n",
      "ClinGen:CA284976,UniProtKB/Swiss-Prot:VAR_064265\n",
      "ClinGen:CA284979,UniProtKB/Swiss-Prot:VAR_064319\n",
      "ClinGen:CA145250,UniProtKB/Swiss-Prot:VAR_043368\n",
      "ClinGen:CA284982,UniProtKB/Swiss-Prot:VAR_064320\n",
      "ClinGen:CA145253,UniProtKB/Swiss-Prot:VAR_043369\n",
      "ClinGen:CA284991,UniProtKB/Swiss-Prot:VAR_064269\n",
      "ClinGen:CA145256,UniProtKB/Swiss-Prot:VAR_043370\n",
      "ClinGen:CA285006,UniProtKB/Swiss-Prot:VAR_064327\n",
      "ClinGen:CA285012,UniProtKB/Swiss-Prot:VAR_064328\n",
      "ClinGen:CA285015,UniProtKB/Swiss-Prot:VAR_064329\n",
      "ClinGen:CA285018,UniProtKB/Swiss-Prot:VAR_029720\n",
      "ClinGen:CA285021,UniProtKB/Swiss-Prot:VAR_064275\n",
      "ClinGen:CA285024,UniProtKB/Swiss-Prot:VAR_064345\n",
      "ClinGen:CA285030,UniProtKB/Swiss-Prot:VAR_064351\n",
      "ClinGen:CA219773,UniProtKB/Swiss-Prot:VAR_043371\n",
      "ClinGen:CA266101,UniProtKB/Swiss-Prot:VAR_029675\n",
      "ClinGen:CA266104,UniProtKB/Swiss-Prot:VAR_064306\n",
      "ClinGen:CA285072,UniProtKB/Swiss-Prot:VAR_064247\n",
      "ClinGen:CA273369,UniProtKB/Swiss-Prot:VAR_029678\n",
      "ClinGen:CA285087,UniProtKB/Swiss-Prot:VAR_064248\n",
      "ClinGen:CA285093,UniProtKB/Swiss-Prot:VAR_029681\n",
      "ClinGen:CA269785,UniProtKB/Swiss-Prot:VAR_029682\n",
      "ClinGen:CA285099,UniProtKB/Swiss-Prot:VAR_064252\n",
      "ClinGen:CA285102,UniProtKB/Swiss-Prot:VAR_064253\n",
      "ClinGen:CA285105,UniProtKB/Swiss-Prot:VAR_029684\n",
      "ClinGen:CA285111,UniProtKB/Swiss-Prot:VAR_029686\n",
      "ClinGen:CA266107,UniProtKB/Swiss-Prot:VAR_029690\n",
      "ClinGen:CA285129,UniProtKB/Swiss-Prot:VAR_064254\n",
      "ClinGen:CA285141,UniProtKB/Swiss-Prot:VAR_064255\n",
      "ClinGen:CA285150,UniProtKB/Swiss-Prot:VAR_064256\n",
      "ClinGen:CA285162,UniProtKB/Swiss-Prot:VAR_064261\n",
      "ClinGen:CA285165,UniProtKB/Swiss-Prot:VAR_029701\n",
      "ClinGen:CA285180,UniProtKB/Swiss-Prot:VAR_064266\n",
      "ClinGen:CA285189,UniProtKB/Swiss-Prot:VAR_029708\n",
      "ClinGen:CA266117,UniProtKB/Swiss-Prot:VAR_029709\n",
      "ClinGen:CA285192,UniProtKB/Swiss-Prot:VAR_064270\n",
      "ClinGen:CA266120,UniProtKB/Swiss-Prot:VAR_029715\n",
      "ClinGen:CA285207,UniProtKB/Swiss-Prot:VAR_029716\n",
      "ClinGen:CA285234,UniProtKB/Swiss-Prot:VAR_064274\n",
      "ClinGen:CA285237,UniProtKB/Swiss-Prot:VAR_029723\n",
      "ClinGen:CA206668,UniProtKB/Swiss-Prot:VAR_029724\n",
      "ClinGen:CA266129,UniProtKB/Swiss-Prot:VAR_029727\n",
      "ClinGen:CA285249,UniProtKB/Swiss-Prot:VAR_029729\n",
      "ClinGen:CA266187,UniProtKB/Swiss-Prot:VAR_009603\n"
     ]
    }
   ],
   "source": [
    "for idx in df_total.index:\n",
    "    if 'UniProt' in df_total.loc[idx,'OtherIDs'] and df_total.loc[idx,'UniPort']=='0':\n",
    "        df_total.loc[idx,'UniPort']=extract_uniprot(df_total.loc[idx,'OtherIDs'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_182771/2711039865.py:71: DtypeWarning: Columns (18) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  f=pd.read_csv(os.path.join(data_path,'clinvar','variant_summary_2022-02.txt'),sep='\\t')\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "def if_positive_or_negative(string_list):\n",
    "    label=[]\n",
    "    for string in string_list:\n",
    "        if string in ['Pathogenic', \n",
    "                    'Pathogenic/Likely pathogenic', \n",
    "                    'probable-pathogenic',\n",
    "                    'Likely pathogenic', \n",
    "                    'pathologic', \n",
    "                    'pathogenic',\n",
    "                    'likely pathogenic',\n",
    "                    'Pathogenic/Likely pathogenic/Established risk allele',\n",
    "                    'likely pathogenic - adrenal pheochromocytoma',\n",
    "                    'Pathogenic/Pathogenic, low penetrance',\n",
    "                    'Pathogenic, low penetrance']:\n",
    "            label.append(1)\n",
    "        elif string in ['Benign',\n",
    "                    'Likely benign',\n",
    "                    'Likely Benign',\n",
    "                    'Benign/Likely benign',\n",
    "                    'non-pathogenic', \n",
    "                    'benign', 'probable-non-pathogenic', 'Likely Benign', 'probably not pathogenic',\n",
    "        ]:\n",
    "            label.append(-1)\n",
    "\n",
    "        else: label.append(0)\n",
    "    return label\n",
    "\n",
    "\n",
    "\n",
    "import requests\n",
    "import json\n",
    "import re\n",
    "def find_fasta_refseq(refseqID):\n",
    "    url='https://rest.uniprot.org/uniprotkb/search?query=%s'%refseqID\n",
    "    cnt=json.loads(requests.get(url).text)\n",
    "    try:\n",
    "        assert re.search('uniprot',cnt.get(\"results\")[0].get(\"entryType\"),re.IGNORECASE)\n",
    "        return cnt[\"results\"][0]['primaryAccession']\n",
    "    except:\n",
    "        print('Error found in '+url)\n",
    "        return 'Error found in '+url\n",
    "\n",
    "def get_uniprot_from_name(name):\n",
    "    try:\n",
    "        refseq=re.match(r'\\S*?(?=[\\(:])',name).group(0)\n",
    "        uniprot=find_fasta_refseq(refseq)\n",
    "\n",
    "    except:\n",
    "        refseq='Error Finding Refseq'\n",
    "        print('Error Finding Refseq')\n",
    "        uniprot=find_fasta_refseq(refseq)\n",
    "    return uniprot \n",
    "\n",
    "def exlude_synonymous(name):\n",
    "    #\"NM_017813.5(BPNT2):c.228C>A (p.Arg76=)\"\n",
    "    return bool(re.match(r'.*(\\(p\\.[a-zA-Z0-9]*=\\)).*',name))\n",
    "\n",
    "def get_refseq_from_name(name):\n",
    "    try:\n",
    "        refseq=re.match(r'\\S*?(?=[\\(:])',name).group(0)\n",
    "\n",
    "    except:\n",
    "        refseq='Error Finding Refseq'\n",
    "        print('Error Finding Refseq')\n",
    "    return refseq \n",
    "\n",
    "\n",
    "data_path='/home/grads/z/zshuying/Documents/shuying/ppi_mutation/data'\n",
    "f=pd.read_csv(os.path.join(data_path,'clinvar','variant_summary_2022-02.txt'),sep='\\t')\n",
    "\n",
    "f['label']=if_positive_or_negative(f['ClinicalSignificance'])\n",
    "\n",
    "keep_cols=['#AlleleID', 'Type', 'Name','label', 'ClinicalSignificance', 'ClinSigSimple', 'ReviewStatus', 'OtherIDs', 'LastEvaluated', 'RS# (dbSNP)',\n",
    "    'nsv/esv (dbVar)', 'RCVaccession']\n",
    "keep_conditions=(f['Type']=='single nucleotide variant') &\\\n",
    "            (['no assertion' not in item for item in f['ReviewStatus'].tolist()]) &\\\n",
    "            (['p' in item for item in f['Name'].tolist()]) &\\\n",
    "            (f['label']!=0) &\\\n",
    "            ([bool(exlude_synonymous(name))==False for name in f['Name'].tolist()])\n",
    "f_simple=f[keep_conditions][keep_cols]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "keep_conditions=(f['Type']=='single nucleotide variant') &\\\n",
    "            (['no assertion' not in item for item in f['ReviewStatus'].tolist()]) &\\\n",
    "            (['p' in item for item in f['Name'].tolist()]) &\\\n",
    "            (f['label']!=0) &\\\n",
    "            ([bool(exlude_synonymous(name))==False for name in f['Name'].tolist()])\n",
    "f_simple=f[keep_conditions][keep_cols]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "shuying",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "ea132143862a5a1585a78ca35dc11c6c8e0adb2bc2e6416d63c557466f5bb992"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
