{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "file = '/home/grads/z/zshuying/Documents/shuying/ppi_mutation/data/clinvar/criteria_above.txt'\n",
    "out='/home/grads/z/zshuying/Documents/shuying/ppi_mutation/data/clinvar/criteria_above_discard_redundant_clinsig.txt'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{' other'}  is discarded\n",
      "{' other'}  is discarded\n",
      "{' other'}  is discarded\n",
      "{' other'}  is discarded\n",
      "{' other'}  is discarded\n",
      "{' other'}  is discarded\n",
      "{' other'}  is discarded\n",
      "{' other'}  is discarded\n",
      "{' risk factor'}  is discarded\n",
      "{' risk factor'}  is discarded\n",
      "{' risk factor'}  is discarded\n",
      "{' other'}  is discarded\n",
      "{' risk factor'}  is discarded\n",
      "{' risk factor'}  is discarded\n",
      "{' other'}  is discarded\n",
      "{' other'}  is discarded\n",
      "{' other'}  is discarded\n",
      "{' other'}  is discarded\n",
      "{' risk factor'}  is discarded\n",
      "{' risk factor', ' association'}  is discarded\n",
      "{' association'}  is discarded\n",
      "{' other'}  is discarded\n",
      "{' other'}  is discarded\n",
      "{' risk factor'}  is discarded\n",
      "{' risk factor'}  is discarded\n",
      "{' other'}  is discarded\n",
      "{' other'}  is discarded\n",
      "{' other'}  is discarded\n",
      "{' other'}  is discarded\n",
      "{' risk factor'}  is discarded\n",
      "{' risk factor'}  is discarded\n",
      "{' other'}  is discarded\n",
      "{' risk factor'}  is discarded\n",
      "{' association'}  is discarded\n",
      "{' risk factor', ' other'}  is discarded\n",
      "{' other'}  is discarded\n",
      "{' other'}  is discarded\n",
      "{' other'}  is discarded\n",
      "{' other'}  is discarded\n",
      "{' other'}  is discarded\n",
      "{' risk factor'}  is discarded\n",
      "{' other'}  is discarded\n",
      "{' other'}  is discarded\n",
      "{' other'}  is discarded\n",
      "{' other'}  is discarded\n",
      "{' other'}  is discarded\n",
      "{' other'}  is discarded\n",
      "{' other'}  is discarded\n",
      "{' other'}  is discarded\n",
      "{' other'}  is discarded\n",
      "{' other'}  is discarded\n",
      "{' risk factor'}  is discarded\n",
      "{' other'}  is discarded\n",
      "{' other'}  is discarded\n",
      "{' risk factor'}  is discarded\n",
      "{' risk factor'}  is discarded\n",
      "{' risk factor'}  is discarded\n",
      "{' association'}  is discarded\n",
      "{' risk factor'}  is discarded\n"
     ]
    }
   ],
   "source": [
    "with open(file,'r') as f:\n",
    "    with open(out,'w') as f_out:\n",
    "        line=f.readline()\n",
    "        f_out.write('clinvar_id;review_status;clinical_sig;uniprot_kb;variant_type;hgvs_p;missense\\n')\n",
    "        while True:\n",
    "            line=f.readline()\n",
    "\n",
    "            if line.count(';')>6:\n",
    "                discard_num=line.count(';')-6+3\n",
    "                line_new=';'.join(line.split(';')[:3]+line.split(';')[discard_num:])\n",
    "                print(set(line.split(';'))-set(line_new.split(';')),' is discarded')\n",
    "                f_out.write(line_new)\n",
    "            else:\n",
    "                f_out.write(line)\n",
    "            if not line:break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>clinvar_id</th>\n",
       "      <th>review_status</th>\n",
       "      <th>clinical_sig</th>\n",
       "      <th>uniprot_kb</th>\n",
       "      <th>variant_type</th>\n",
       "      <th>hgvs_p</th>\n",
       "      <th>missense</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>RCV000000012</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Q16570#VAR_015068</td>\n",
       "      <td>Variation</td>\n",
       "      <td>LRG_801p1:p.Trp96fs</td>\n",
       "      <td>frameshift variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>RCV000000070</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>O95876#VAR_064770</td>\n",
       "      <td>Variation</td>\n",
       "      <td>O95876:p.Arg55Lys</td>\n",
       "      <td>frameshift variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>RCV000000113</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>P06132#VAR_007914</td>\n",
       "      <td>Variation</td>\n",
       "      <td>P06132:p.Asn304Lys</td>\n",
       "      <td>missense variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>RCV000000178</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>P04181#VAR_000577</td>\n",
       "      <td>Variation</td>\n",
       "      <td>P04181:p.Ala270Pro</td>\n",
       "      <td>missense variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SCV002232601</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>P04181#VAR_000578</td>\n",
       "      <td>Variation</td>\n",
       "      <td>P04181:p.Arg271Lys</td>\n",
       "      <td>missense variant</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     clinvar_id                        review_status clinical_sig  \\\n",
       "0  RCV000000012  criteria provided, single submitter   Pathogenic   \n",
       "1  RCV000000070  criteria provided, single submitter   Pathogenic   \n",
       "2  RCV000000113  criteria provided, single submitter   Pathogenic   \n",
       "3  RCV000000178  criteria provided, single submitter   Pathogenic   \n",
       "4  SCV002232601  criteria provided, single submitter   Pathogenic   \n",
       "\n",
       "          uniprot_kb variant_type               hgvs_p            missense  \n",
       "0  Q16570#VAR_015068    Variation  LRG_801p1:p.Trp96fs  frameshift variant  \n",
       "1  O95876#VAR_064770    Variation    O95876:p.Arg55Lys  frameshift variant  \n",
       "2  P06132#VAR_007914    Variation   P06132:p.Asn304Lys    missense variant  \n",
       "3  P04181#VAR_000577    Variation   P04181:p.Ala270Pro    missense variant  \n",
       "4  P04181#VAR_000578    Variation   P04181:p.Arg271Lys    missense variant  "
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data=pd.read_csv(out,delimiter=';')\n",
    "data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Pathogenic', 'Uncertain significance',\n",
       "       'Pathogenic/Likely pathogenic', 'probable-pathogenic',\n",
       "       'Likely pathogenic', 'risk factor', 'pathologic', 'Benign',\n",
       "       'not provided', 'Affects', 'association',\n",
       "       'Conflicting interpretations of pathogenicity', 'Likely benign',\n",
       "       'protective', 'other', 'drug response', 'pathogenic',\n",
       "       'variant of unknown significance', 'Benign/Likely benign',\n",
       "       'unknown', 'untested', 'non-pathogenic', 'Likely risk allele',\n",
       "       'benign', 'probable-non-pathogenic', 'Likely Benign', 'Unknown',\n",
       "       'Uncertain risk allele', 'confers sensitivity',\n",
       "       'likely pathogenic',\n",
       "       'Pathogenic/Likely pathogenic/Established risk allele',\n",
       "       'no known pathogenicity', 'probably not pathogenic',\n",
       "       'Unknown significance',\n",
       "       'likely pathogenic - adrenal pheochromocytoma',\n",
       "       'Pathogenic/Pathogenic, low penetrance'], dtype=object)"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data.clinical_sig.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Variation', 'single nucleotide variant', 'copy number gain',\n",
       "       'Deletion', 'Duplication', 'copy number loss', 'Complex',\n",
       "       'Insertion', 'Translocation', 'Indel', 'Tandem duplication'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data.variant_type.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>clinvar_id</th>\n",
       "      <th>review_status</th>\n",
       "      <th>clinical_sig</th>\n",
       "      <th>uniprot_kb</th>\n",
       "      <th>variant_type</th>\n",
       "      <th>hgvs_p</th>\n",
       "      <th>missense</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>850</th>\n",
       "      <td>RCV000082831</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>O75445#VAR_072020</td>\n",
       "      <td>Deletion</td>\n",
       "      <td>O75445:p.Ser2639Pro</td>\n",
       "      <td>missense variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>917</th>\n",
       "      <td>RCV000085540</td>\n",
       "      <td>criteria provided, multiple submitters, no con...</td>\n",
       "      <td>Pathogenic/Likely pathogenic</td>\n",
       "      <td>P78363#VAR_012552</td>\n",
       "      <td>Deletion</td>\n",
       "      <td>P78363:p.Leu1014Arg</td>\n",
       "      <td>missense variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1196</th>\n",
       "      <td>RCV000174262</td>\n",
       "      <td>criteria provided, multiple submitters, no con...</td>\n",
       "      <td>Benign</td>\n",
       "      <td>P08183#VAR_015003</td>\n",
       "      <td>Deletion</td>\n",
       "      <td>P08183:p.Ser400Asn</td>\n",
       "      <td>missense variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1202</th>\n",
       "      <td>RCV000175792</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>Q14003#VAR_074198</td>\n",
       "      <td>Deletion</td>\n",
       "      <td>Q14003:p.Asp477Asn</td>\n",
       "      <td>missense variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1266</th>\n",
       "      <td>RCV000190521</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>Likely pathogenic</td>\n",
       "      <td>Q14520#VAR_023401</td>\n",
       "      <td>Deletion</td>\n",
       "      <td>Q13426:p.Trp43Arg</td>\n",
       "      <td>non-coding transcript variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44185</th>\n",
       "      <td>RCV001711641</td>\n",
       "      <td>criteria provided, multiple submitters, no con...</td>\n",
       "      <td>Benign</td>\n",
       "      <td>Q96Q42#VAR_036748</td>\n",
       "      <td>Deletion</td>\n",
       "      <td>Q96Q42:p.Glu159Lys</td>\n",
       "      <td>missense variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44391</th>\n",
       "      <td>RCV001770045</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>P07949#VAR_009482</td>\n",
       "      <td>Deletion</td>\n",
       "      <td>P07949:p.Leu790Phe</td>\n",
       "      <td>missense variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44624</th>\n",
       "      <td>RCV001787847</td>\n",
       "      <td>reviewed by expert panel</td>\n",
       "      <td>drug response</td>\n",
       "      <td>P21817#VAR_045728</td>\n",
       "      <td>Deletion</td>\n",
       "      <td>NP_000531.2:p.R2452W</td>\n",
       "      <td>missense variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45804</th>\n",
       "      <td>RCV002225273</td>\n",
       "      <td>criteria provided, multiple submitters, no con...</td>\n",
       "      <td>Benign/Likely benign</td>\n",
       "      <td>P04637#VAR_014632</td>\n",
       "      <td>Deletion</td>\n",
       "      <td>P04637:p.Pro47Ser</td>\n",
       "      <td>missense variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46139</th>\n",
       "      <td>RCV002259313</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>Likely pathogenic</td>\n",
       "      <td>P51787#VAR_074946</td>\n",
       "      <td>Deletion</td>\n",
       "      <td>LRG_287p2:p.Arg63Trp</td>\n",
       "      <td>missense variant</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>171 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         clinvar_id                                      review_status  \\\n",
       "850    RCV000082831                criteria provided, single submitter   \n",
       "917    RCV000085540  criteria provided, multiple submitters, no con...   \n",
       "1196   RCV000174262  criteria provided, multiple submitters, no con...   \n",
       "1202   RCV000175792                criteria provided, single submitter   \n",
       "1266   RCV000190521                criteria provided, single submitter   \n",
       "...             ...                                                ...   \n",
       "44185  RCV001711641  criteria provided, multiple submitters, no con...   \n",
       "44391  RCV001770045                criteria provided, single submitter   \n",
       "44624  RCV001787847                           reviewed by expert panel   \n",
       "45804  RCV002225273  criteria provided, multiple submitters, no con...   \n",
       "46139  RCV002259313                criteria provided, single submitter   \n",
       "\n",
       "                       clinical_sig         uniprot_kb variant_type  \\\n",
       "850          Uncertain significance  O75445#VAR_072020     Deletion   \n",
       "917    Pathogenic/Likely pathogenic  P78363#VAR_012552     Deletion   \n",
       "1196                         Benign  P08183#VAR_015003     Deletion   \n",
       "1202         Uncertain significance  Q14003#VAR_074198     Deletion   \n",
       "1266              Likely pathogenic  Q14520#VAR_023401     Deletion   \n",
       "...                             ...                ...          ...   \n",
       "44185                        Benign  Q96Q42#VAR_036748     Deletion   \n",
       "44391                    Pathogenic  P07949#VAR_009482     Deletion   \n",
       "44624                 drug response  P21817#VAR_045728     Deletion   \n",
       "45804          Benign/Likely benign  P04637#VAR_014632     Deletion   \n",
       "46139             Likely pathogenic  P51787#VAR_074946     Deletion   \n",
       "\n",
       "                     hgvs_p                       missense  \n",
       "850     O75445:p.Ser2639Pro               missense variant  \n",
       "917     P78363:p.Leu1014Arg               missense variant  \n",
       "1196     P08183:p.Ser400Asn               missense variant  \n",
       "1202     Q14003:p.Asp477Asn               missense variant  \n",
       "1266      Q13426:p.Trp43Arg  non-coding transcript variant  \n",
       "...                     ...                            ...  \n",
       "44185    Q96Q42:p.Glu159Lys               missense variant  \n",
       "44391    P07949:p.Leu790Phe               missense variant  \n",
       "44624  NP_000531.2:p.R2452W               missense variant  \n",
       "45804     P04637:p.Pro47Ser               missense variant  \n",
       "46139  LRG_287p2:p.Arg63Trp               missense variant  \n",
       "\n",
       "[171 rows x 7 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data[data.variant_type=='Deletion']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>clinvar_id</th>\n",
       "      <th>review_status</th>\n",
       "      <th>clinical_sig</th>\n",
       "      <th>uniprot_kb</th>\n",
       "      <th>variant_type</th>\n",
       "      <th>hgvs_p</th>\n",
       "      <th>missense</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1117</th>\n",
       "      <td>RCV000153954</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>P36021#VAR_074574</td>\n",
       "      <td>Duplication</td>\n",
       "      <td>P36021:p.Arg197His</td>\n",
       "      <td>missense variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1204</th>\n",
       "      <td>RCV000176186</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>P02458#VAR_063895</td>\n",
       "      <td>Duplication</td>\n",
       "      <td>P02458:p.Gly453Ala</td>\n",
       "      <td>missense variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1765</th>\n",
       "      <td>RCV000359374</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>Likely pathogenic</td>\n",
       "      <td>P08236#VAR_037922</td>\n",
       "      <td>Duplication</td>\n",
       "      <td>P08236:p.Tyr320Ser</td>\n",
       "      <td>non-coding transcript variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1986</th>\n",
       "      <td>RCV000518763</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>P30793#VAR_016907</td>\n",
       "      <td>Duplication</td>\n",
       "      <td>P30793:p.Arg249Ser</td>\n",
       "      <td>missense variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2557</th>\n",
       "      <td>RCV001547695</td>\n",
       "      <td>criteria provided, multiple submitters, no con...</td>\n",
       "      <td>Pathogenic/Likely pathogenic</td>\n",
       "      <td>Q14807#VAR_067347</td>\n",
       "      <td>Duplication</td>\n",
       "      <td>Q14807:p.Arg149Leu</td>\n",
       "      <td>missense variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43589</th>\n",
       "      <td>RCV001556734</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>P51787#VAR_074937</td>\n",
       "      <td>Duplication</td>\n",
       "      <td>LRG_287p2:p.Glu19Lys</td>\n",
       "      <td>missense variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43670</th>\n",
       "      <td>RCV001573696</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>Benign</td>\n",
       "      <td>Q7Z4L5#VAR_065521</td>\n",
       "      <td>Duplication</td>\n",
       "      <td>Q7Z4L5:p.Asp242Asn</td>\n",
       "      <td>missense variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43788</th>\n",
       "      <td>RCV001618392</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>Benign</td>\n",
       "      <td>P01222#VAR_054769</td>\n",
       "      <td>Duplication</td>\n",
       "      <td>P01222:p.Thr14Ala</td>\n",
       "      <td>missense variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43792</th>\n",
       "      <td>RCV001636605</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>Benign</td>\n",
       "      <td>P01009#VAR_006991</td>\n",
       "      <td>Duplication</td>\n",
       "      <td>P01009:p.Arg125His</td>\n",
       "      <td>missense variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44520</th>\n",
       "      <td>RCV001787706</td>\n",
       "      <td>reviewed by expert panel</td>\n",
       "      <td>drug response</td>\n",
       "      <td>P21817#VAR_005590</td>\n",
       "      <td>Duplication</td>\n",
       "      <td>LRG_766p1:p.Arg163Cys</td>\n",
       "      <td>missense variant</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>61 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         clinvar_id                                      review_status  \\\n",
       "1117   RCV000153954                criteria provided, single submitter   \n",
       "1204   RCV000176186                criteria provided, single submitter   \n",
       "1765   RCV000359374                criteria provided, single submitter   \n",
       "1986   RCV000518763                criteria provided, single submitter   \n",
       "2557   RCV001547695  criteria provided, multiple submitters, no con...   \n",
       "...             ...                                                ...   \n",
       "43589  RCV001556734                criteria provided, single submitter   \n",
       "43670  RCV001573696                criteria provided, single submitter   \n",
       "43788  RCV001618392                criteria provided, single submitter   \n",
       "43792  RCV001636605                criteria provided, single submitter   \n",
       "44520  RCV001787706                           reviewed by expert panel   \n",
       "\n",
       "                       clinical_sig         uniprot_kb variant_type  \\\n",
       "1117         Uncertain significance  P36021#VAR_074574  Duplication   \n",
       "1204                     Pathogenic  P02458#VAR_063895  Duplication   \n",
       "1765              Likely pathogenic  P08236#VAR_037922  Duplication   \n",
       "1986         Uncertain significance  P30793#VAR_016907  Duplication   \n",
       "2557   Pathogenic/Likely pathogenic  Q14807#VAR_067347  Duplication   \n",
       "...                             ...                ...          ...   \n",
       "43589        Uncertain significance  P51787#VAR_074937  Duplication   \n",
       "43670                        Benign  Q7Z4L5#VAR_065521  Duplication   \n",
       "43788                        Benign  P01222#VAR_054769  Duplication   \n",
       "43792                        Benign  P01009#VAR_006991  Duplication   \n",
       "44520                 drug response  P21817#VAR_005590  Duplication   \n",
       "\n",
       "                      hgvs_p                       missense  \n",
       "1117      P36021:p.Arg197His               missense variant  \n",
       "1204      P02458:p.Gly453Ala               missense variant  \n",
       "1765      P08236:p.Tyr320Ser  non-coding transcript variant  \n",
       "1986      P30793:p.Arg249Ser               missense variant  \n",
       "2557      Q14807:p.Arg149Leu               missense variant  \n",
       "...                      ...                            ...  \n",
       "43589   LRG_287p2:p.Glu19Lys               missense variant  \n",
       "43670     Q7Z4L5:p.Asp242Asn               missense variant  \n",
       "43788      P01222:p.Thr14Ala               missense variant  \n",
       "43792     P01009:p.Arg125His               missense variant  \n",
       "44520  LRG_766p1:p.Arg163Cys               missense variant  \n",
       "\n",
       "[61 rows x 7 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data[data.variant_type=='Duplication']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>clinvar_id</th>\n",
       "      <th>review_status</th>\n",
       "      <th>clinical_sig</th>\n",
       "      <th>uniprot_kb</th>\n",
       "      <th>variant_type</th>\n",
       "      <th>hgvs_p</th>\n",
       "      <th>missense</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1361</th>\n",
       "      <td>RCV000207037</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>O00754#VAR_068061</td>\n",
       "      <td>Complex</td>\n",
       "      <td>O00754:p.Gly800Trp</td>\n",
       "      <td>synonymous variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2812</th>\n",
       "      <td>RCV001844011</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>P10828#VAR_004634</td>\n",
       "      <td>Complex</td>\n",
       "      <td>P10828:p.Arg316His</td>\n",
       "      <td>missense variant</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        clinvar_id                        review_status  \\\n",
       "1361  RCV000207037  criteria provided, single submitter   \n",
       "2812  RCV001844011  criteria provided, single submitter   \n",
       "\n",
       "                clinical_sig         uniprot_kb variant_type  \\\n",
       "1361  Uncertain significance  O00754#VAR_068061      Complex   \n",
       "2812              Pathogenic  P10828#VAR_004634      Complex   \n",
       "\n",
       "                  hgvs_p            missense  \n",
       "1361  O00754:p.Gly800Trp  synonymous variant  \n",
       "2812  P10828:p.Arg316His    missense variant  "
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data[data.variant_type=='Complex']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "missense_data=data[data.missense=='missense variant']"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Process clinical significance"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "what is protective?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "def if_positive_or_negative(string):\n",
    "    if string in ['Pathogenic', \n",
    "                'Pathogenic/Likely pathogenic', \n",
    "                'probable-pathogenic',\n",
    "                'Likely pathogenic', \n",
    "                'pathologic', \n",
    "                'pathogenic',\n",
    "                'likely pathogenic',\n",
    "                'Pathogenic/Likely pathogenic/Established risk allele',\n",
    "                'likely pathogenic - adrenal pheochromocytoma',\n",
    "                'Pathogenic/Pathogenic, low penetrance']:\n",
    "        return 1\n",
    "    elif string in ['Benign',\n",
    "                'Likely benign',\n",
    "                'Benign/Likely benign',\n",
    "                'non-pathogenic', \n",
    "                'benign', 'probable-non-pathogenic', 'Likely Benign', 'probably not pathogenic',\n",
    "    ]:\n",
    "        return -1\n",
    "\n",
    "   else: return 0\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "missense_data['label']='1'"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "shuying",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "ea132143862a5a1585a78ca35dc11c6c8e0adb2bc2e6416d63c557466f5bb992"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
